ssl: Add hardening guide (#11019)
erlang/otp: lib/ssl/doc/guides/ssl_hardening.md

You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

%CopyrightEnd%
-->

# TLS Harding Guide

## Introduction
There can exist trade-offs between interoperability and security, and
between resource consumption and security. If you make such
trade-offs it is important that you make an informed decision and
understand the consequences.

Erlang's TLS implementation provided by the `ssl` application strives to
have secure defaults. That said, we will avoid changing defaults in
non-major releases.

## Cryptographic Algorithms
The available cryptographic algorithms depend on the version of
OpenSSL cryptolib that Erlang/OTP is built and linked with.

The configuration algorithm lists will be automatically filtered
except for cipher suites (`ciphers` option)
that pre TLS-1.3 combines several algorithms

> and you will need to call `ssl:filter_cipher_suites/2`

Reviewer: ### 5.23 Sentences missing termination period

**Severity:** Minor
**WHY:** Three paragraph-ending lines lack a termination period.
**WHAT:**

| Line | Text | Addressed by bjorng? |
|------|------|---------------------|
| 44 | `and you will need to call ssl:filter_cipher_suites/2` | Orphaned (wrong path) — must fix manually |
| 58 | `Only supported in the latest protocol version TLS-1.3` | ✅ Yes (comment #8, actionable) |
| 103 | `the key exchange` | ✅ Yes (comment #14, actionable) |

**HOW:** Line 44 needs a manual fix (append `.`). Lines 58 and 103
will be fixed if Ingela accepts bjorng's suggestions.

Author: I think this is not relevant any more.

than the previous renegotiation mechanism.

Client renegotiation can be disabled and is by default mitigate
by the server not allowing two renegotiations close to each other.

### Key Exchange Groups
TLS-1.3 decouples key exchange algorithms form cipher suites. These
are configured using the `groups` option. In TLS-1.2 in addition to
the cipher suites the options `eccs`, `dh`, `dhfile` exist to configure
the key exchange

### Signature Algorithms
The options `signature_algs` and `signature_algs_cert` can be used to
allow different signature algorithms for protocol messages and
certificates. If the latter is not specified it will default to
former.

> #### Note {: .info }
Currently the default will add `signature_algs_cert` that allows
signature algorithms using sha1 as pseudo random function for
certificate signatures, which is not allowed by default for protocol
signatures.  Expected to be removed after year 2030 when such
certificates should have been phased out.

### Certificate and Keys
Use the `certs_keys` option to be able to configure more than one
possible certificate for the client or server. The `best`
one compatible with the peer will be chosen.

### Pre Shared Keys

> Do not use the same PSK across both TLS 1.2 and TLS 1.3, as it is

Reviewer: ### 5.29 Acronyms not expanded on first use

**Severity:** Minor
**WHY:** Several acronyms are used without expansion. CRL, OCSP, and
NSS have no expansion anywhere in the guide — a reader unfamiliar with
these terms has no way to look them up from context. PSK is implicitly
connected to its heading "Pre Shared Keys" but lacks the standard
"(PSK)" parenthetical and a hyphen. Not covered by bjorng.
**WHAT:**

| Acronym | First use | Expand? |
|---------|-----------|---------|
| CRL | Line 150 | Yes — Certificate Revocation List (no expansion anywhere) |
| OCSP | Line 152 | Yes — Online Certificate Status Protocol (no expansion anywhere) |
| NSS | Line 68 | Yes — Network Security Services (no expansion anywhere) |
| PSK | Line 119 | Nit — heading "Pre Shared Keys" makes it guessable, but add "(PSK)" and hyphen: "Pre-Shared Keys (PSK)" per RFC 4279 |
| TLS, DTLS, UDP, AES, CCM | Various | No — audience for a TLS hardening guide knows these |

**HOW:** Expand on first use, e.g.:
- Line 150: `Certificate Revocation List (CRL) check option...`
- Line 152: `...support OCSP (Online Certificate Status Protocol) stapling...`
- Line 68: `### NSS (Network Security Services) Key Logging`
- Heading: `### Pre-Shared Keys (PSK)`

Author: fixed

> considered unsafe.

Reviewer: ### 5.30 Security warnings missing admonition blocks

**Severity:** Minor
**WHY:** The guide has 4 sentences warning about unsafe/insecure
practices, all rendered as plain text. The guide already uses
`> #### Note {: .info }` blocks for informational notes (lines 46,
111) but has zero `> #### Warning {: .warning }` blocks. Security
warnings should be visually distinct. Not covered by bjorng.
**WHAT:**

| Line | Warning text | Recommended block |
|------|-------------|-------------------|
| 61–62 | "enabling anything documented as legacy will make your application less secure" | `{: .warning }` |
| 69–70 | NSS key logging "should not be enabled in production...defeats the purpose of TLS" | `{: .warning }` |
| 119 | "Do not use the same PSK across both TLS 1.2 and TLS 1.3...unsafe" | `{: .warning }` |
| 139–140 | verify_fun "at own risk and will weaken the security properties" | `{: .warning }` |

**HOW:** Wrap each in a warning admonition, e.g.:
```markdown
> #### Warning {: .warning }
> Do not use the same PSK across both TLS 1.2 and TLS 1.3,
> as it is considered unsafe.
```

Author: fixed

> In TLS-1.2 PSK is supported by PSK cipher suites and generally lack

Reviewer: ### 5.31 PSK paragraph grammar and precision

**Severity:** Minor
**WHY:** Lines 121–123 have multiple grammar errors and a misleading
technical claim. Not covered by bjorng.
**WHAT:**

Line 121: `In TLS-1.2 PSK is supported by PSK cipher suites and generally lack forward secrecy.`
- "lack" → "lacks" (singular subject)
- Missing comma after "TLS-1.2"
- "generally lack forward secrecy" is misleading — plain `psk` suites
  lack it, but `dhe_psk` and `ecdhe_psk` provide forward secrecy

Line 122–123: `In TLS-1.3  PSK can be use together with a key exchange algorithm and then provides forwar\
d secrecy.`
- "be use" → "be used"
- Double space after "TLS-1.3"
- Missing comma after "TLS-1.3"

**HOW:** Suggested rewrite:

> In TLS-1.2, PSK is supported by PSK cipher suites, which generally
> lack forward secrecy unless combined with DHE or ECDHE key exchange.
> In TLS-1.3, PSK can be used together with a key exchange algorithm,
> providing forward secrecy.

Author: Rephrased with inspiration from your suggestion.

forward secrecy.  In TLS-1.3 PSK can be use together with a key
exchange algorithm and then provides forward secrecy.

### Early Data

> The security properties of Early Data introduced in TLS-1.3 are weaker

Reviewer: ### 4.8 (AP-31) Early Data Section

**Severity:** Minor
**WHY:** Lines 157-161 read:

> The security properties of Early Data introduced in TLS-1.3 are
> weaker than other kinds of TLS data. See [Early Data in
> TLS-1.3](using_ssl.html#early-data-in-tls-1-3) for more information
> of what to consider.

Four problems:

1. **Link format: `.html` → `.md`** — OTP ExDoc links should use
   `using_ssl.md#early-data-in-tls-1-3`.

2. **"for more information of what" → "on what"** — "information of"
   is not idiomatic English.

3. **"weaker" is unhelpfully vague for a hardening guide.** The linked
   `using_ssl` guide spells out the specific risks: early data is
   **not forward secret** and is **vulnerable to replay attacks**. A
   hardening guide should state these risks inline — the reader
   shouldn't have to click through to learn the security implications.

4. **No mention of the `early_data` option or its default.** The
   server option `early_data` controls whether the server accepts
   early data (default: `disabled`). A hardening guide should state
   that the default is secure and warn about enabling it.

**WHAT:** `ssl_hardening.md`, lines 157-161.
**HOW:** Suggested rewrite:

> Early data (0-RTT) in TLS-1.3 is not forward secret and is
> vulnerable to replay attacks. The server option
> [`early_data`](`t:ssl:server_option/0`) is disabled by default.
>
> > #### Warning {: .warning }
> > Only enable early data if your application can safely handle
> > replayed requests (e.g., idempotent HTTP methods).
>
> See [Early Data in TLS-1.3](using_ssl.md#early-data-in-tls-1-3)
> for mitigation strategies and examples.

Author: Added

than other kinds of TLS data. See [Early Data in TLS-1.3](using_ssl.html#early-data-in-tls-1-3) for more information of what to consider.

## Certificate Verification
### Verify Function
The `verify_fun` option to TLS clients and servers lets you customize

> the certificate path validation, that is, validation the chain of

Reviewer: ### 5.32 "validation the chain" — missing "of"

**Severity:** Minor ⚡ bjorng (partially — bjorng rewrites the sentence but this specific error is worth \
noting)
**WHY:** Line 135: "that is, validation the chain of certificates" is
missing "of" — should be "that is, validation of the chain of
certificates". ~~The comma after "that is" is also missing.~~ Comma
fixed in fixup commit (fc4a7906), but "of" still missing.
**WHAT:** Line 135.
**HOW:** Change "validation the chain" to "validation **of** the chain".

Author: fixed

certificates that shall ensure the authenticity of the peer, and its
purpose is to allow for application-specific additional checks.
However, the verify fun can also be used to accept certain

> errors. Doing so is at your own risk and will weaken the security

Reviewer: ### 5.33 "verify fun" inconsistent with "verify_fun"

**Severity:** Nit
**WHY:** Line 138 uses "verify fun" (no underscore, no backticks) while
line 134 uses "verify_fun" (with underscore, no backticks). The option
name is `verify_fun` — all references should use backtick-wrapped
`` `verify_fun` ``.
**WHAT:** Line 138: `The verify fun can however also be used...`
**HOW:** Change to: `` The `verify_fun` can, however, also be used... ``

Author: fixed

properties of your application.

### Certificate Revocation Verification
Certificate revocation verification is important for security and should
not be skipped. However, it needs to be configured as the `ssl`
application does not know how the user's application handles
this. There is a basic default cache that can be externally
populated or rely on certificate HTTP distribution-points. There is
also a callback API that can be used to access an external cache.

The CRL check option `best_effort` is to be considered legacy.

> The client can also support OCSP stapling when connecting to a server

Reviewer: ### 5.38 "support" → "request" OCSP stapling

**Severity:** Minor
**WHY:** Line 153: "The client can also support OCSP stapling" uses
"support" which implies a passive capability. In the TLS protocol, the
client actively *requests* OCSP stapling by including a
`status_request` extension in the ClientHello (RFC 6066). The server
then decides whether to provide the staple. The `ssl` module docs
(`ssl.erl:1504`) use "enabled" for the configuration action. "Request"
is the most accurate verb, matching both RFC terminology and the actual
protocol behavior. Additionally, the qualifier "if connecting to a
server that supports that" is misleading — the client sends the
request regardless; it has no way to know in advance whether the
server supports stapling.
**WHAT:** Line 153: `The client can also support OCSP stapling if
connecting to a server that supports that.`
**HOW:** Change to:

> The client can also request OCSP stapling from the server. In
> TLS-1.2, only end-entity certificate stapling is supported. If the
> server does not provide a staple, the connection will fail with
> `{bad_cert, missing_ocsp_staple}`.

Author: fixed

supporting that. In TLS-1.2, only end entity certificate stapling is supported.

> Client will fail if no staple is provided. Plain OCSP is not supported.

Reviewer: ### 5.39 TLS 1.3 intermediate cert stapling: partial support

**Severity:** Nit (non-blocking, informational)
**WHY:** The guide states "In TLS-1.2 only end entity certificate
staple is supported" (line 154), implying TLS 1.3 supports full chain
stapling. Source code analysis reveals the support is **partial**:

- **Parsing:** `tls_handshake_1_3:split_cert_entries/4` (line 823)
  iterates all CertificateEntry items and extracts `status_request`
  extensions from any cert in the chain — not just the end entity.
- **Validation:** `ssl_certificate:verify_cert_extensions/4` (line 611)
  validates an OCSP response against the cert's issuer using
  `public_key:pkix_ocsp_validate/5` — works for any cert.
- **Enforcement gap:** The stapling state is a single map with one
  `status` field (`received_staple` / `not_received`). It flips to
  `received_staple` on the first staple found. The hard-fail check
  in `ssl_handshake.erl:2379` only verifies that *some* staple was
  received — it does not enforce that every cert in the chain has one.

So if the server provides a staple for the peer cert but not for
intermediates, the connection succeeds. The guide's implication of
full chain support is slightly misleading.

**WHAT:** Line 154.
**HOW:** The current wording is acceptable for a hardening guide — this
is an implementation nuance, not a user-facing configuration issue.
If Ingela wants to be precise:

> In TLS-1.2, only end-entity certificate stapling is supported. In
> TLS-1.3, the server may provide OCSP staples for any certificate in
> the chain, but the client only enforces that the peer certificate
> has a staple.

Author: Humm ... we should look over the implementation and the documentation I think it is lacking. Will adhere to your suggestion for hardening guide for now, it is a acceptable start for something being phased out anyway.

<!--
%CopyrightBegin%

SPDX-License-Identifier: Apache-2.0

Copyright Ericsson AB 2026-2026. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

%CopyrightEnd%
-->

> # TLS Harding Guide

Reviewer: ### 5.1 Title Typo: "Harding" → "Hardening"

**Severity:** Major
**WHY:** The document title reads `# TLS Harding Guide` — missing the "en".
**WHAT:** Line 23 of `ssl_hardening.md`.
**HOW:** Change to `# TLS Hardening Guide`.

Author: fixed

> # TLS Harding Guide

Reviewer: ### 5.40 Section structure assessment

**Severity:** Nit (non-blocking, follow-up PR)
**WHY:** The guide's section hierarchy has three structural issues that
affect logical flow. None prevent a reader from finding information
(the guide is short enough to scan entirely), but they would improve
readability in a follow-up revision.

**Issue 1 — "General Principles" is misplaced and thin.**
The section sits between "Cryptographic Algorithms" and "Protocol
Versions" but contains 5 single-sentence subsections that belong
elsewhere:

| Subsection | Better home |
|------------|-------------|
| Algorithms selection | Cryptographic Algorithms |
| Post Quantum Algorithms | Cryptographic Algorithms or Protocol Versions (TLS-1.3) |
| Legacy | Introduction (general warning) |
| Embedded Systems | Cipher Suites |
| NSS Key Logging | Standalone or Introduction |

**Issue 2 — "Certificate and Keys" is under "Protocol Versions".**
The `certs_keys` option is about certificate configuration, not
protocol version behavior. It belongs under "Certificate Verification"
or as a peer-level `##` section.

**Issue 3 — Mixed abstraction under "Protocol Versions".**
The section mixes protocol-level concepts (version downgrade, DTLS),
configuration topics (cipher suites, key exchange, signature
algorithms), and key management (PSK, certificates). The SSH hardening
guide separates these into distinct top-level sections: DoS →
Verification → Crypto → Protocol → Network.

**Comparison with SSH hardening guide structure:**

| SSH guide flow | SSL guide flow | Notes |
|----------------|----------------|-------|
| Introduction | Introduction | ✅ Both set context |
| DoS resilience | *(missing)* | Acceptable for v1 |
| Verifying remote peer | Certificate Verification | ✅ Equivalent |
| Crypto (algorithms, re-keying) | Crypto Algorithms + Protocol Versions | ⚠️ Mixed |
| Protocol hardening | *(N/A — no shell/exec in SSL)* | N/A |
| Network-level security | *(missing)* | Acceptable for v1 |
| Advanced authentication | *(PSK partially)* | Acceptable for v1 |

**Verdict:** Good enough for an initial version. The SSH hardening
guide also evolved over multiple PRs. Restructuring can happen in a
follow-up without blocking this PR.

**HOW (follow-up):** Suggested restructured outline:

```
# TLS Hardening Guide
  (intro paragraphs directly under title, no ## Introduction)
## Cryptographic Algorithms
  (absorb: algorithm selection, PQC, legacy, embedded, OpenSSL note)
## Protocol Versions
  (keep: version downgrade, DTLS)
## Cipher Suites and Key Exchange
  (move from under Protocol Versions, add groups/eccs/dh)
## Signature Algorithms
## Pre-Shared Keys (PSK)
## Certificates and Keys
  (move from Protocol Versions)
## Certificate Verification
  ### Verify Function
  ### Certificate Revocation (CRL, OCSP)
## Debugging
  (NSS key logging)
```

Author: It has been restructured but not according to the suggestion here. But please lets continue form the new structure. I am not sure that SSH structure fits TLS, the protocols are at different levels.

> ## Introduction

Reviewer: ### 5.41 Remove `## Introduction` heading

**Severity:** Minor
**WHY:** PR-11012 (`84a530ae33`, by u3s) removes `## Introduction`
wrapper headings from the SSH hardening guide and other SSH docs as
part of a heading restructure. The intro content is kept but flows
directly under the `# Hardening` title. The SSL hardening guide should
follow the same pattern for consistency across OTP documentation.
Additionally, `## Introduction` clutters the rendered navigation menu
with a repeated, non-descriptive link.
**WHAT:** Line 25: `## Introduction`
**HOW:** Remove the `## Introduction` line. The 3 intro sentences
remain directly under `# TLS Hardening Guide`. Since PR-11012 targets
`master` and PR-11019 also targets `master`, they should be consistent
at merge time.

Author: fixed

> There can exist trade-offs between interoperability and security, and

Reviewer: ### 5.43 Intro paragraph lacks orientation

**Severity:** Minor
**WHY:** The opening paragraph jumps straight into trade-offs without
telling the reader what this document is or what it covers. The SSH
hardening guide opens with context ("The Erlang/OTP SSH application is
intended to be used...") and closes with purpose ("The purpose of this
guide is to discuss the different hardening options available"). The
SSL guide has no equivalent framing. Additionally, the trade-off
statement is important but buried in plain prose — an `{: .info }`
block (already used elsewhere in the guide) would make it stand out.
**WHAT:** Lines 27–35 (the entire intro content).
**HOW:** Suggested rewrite:

> This guide describes how to harden TLS connections configured with
> the `ssl` application. It covers cryptographic algorithm selection,
> protocol version configuration, certificate verification, and
> revocation checking.
>
> > #### Note {: .info }
> > There can be trade-offs between interoperability and security, and
> > between resource consumption and security. If you make such
> > trade-offs, it is important that you make an informed decision and
> > understand the consequences.
>
> The `ssl` application strives to have secure defaults. Defaults are
> only changed in major releases.

Author: Adjust to proposal.  I do not want to say only ssl application as I want to emphasis that the protocol we are talking about is called TLS although we are for practical reasons not renaming the application.  Also I do not want to say that we only change defaults in major release as for a good enough reason we might break that rule.

Reviewer: should ssl vs tls naming be explained in the document opening section? info block maybe?

between resource consumption and security. If you makes such
trade-offs it is important that you make an informed decision and
understand the consequences.

Erlang TLS implementation provided by the ssl application strives to
have secure defaults, that said we will avoid changing defaults in
none major releases.

## Cryptographic Algorithms

The available cryptographic algorithms depends on the version of
OpenSSL cryptolib that Erlang/OTP is built and linked with.

Configuration algorithm lists will be automatically filtered
except for cipher suites configured by the `ciphers` option
that pre TLS-1.3 combines several algorithms
and you will need to call ssl:filter_cipher_suites/2

> #### Note {: .info }

> Note that the TLS protocol is implemented in Erlang so only

Reviewer: ### 5.44 Redundant "Note that" inside Note block

**Severity:** Nit
**WHY:** Line 47 starts with "Note that" inside a `> #### Note {: .info }`
admonition block. The block header already signals "this is a note" —
repeating it in the body is redundant. Not covered by bjorng (his
orphaned comment #6 rewrites the sentence but keeps "Note that").
**WHAT:** Line 47: `Note that the TLS protocol is implemented in Erlang so only`
**HOW:** Change to: `The TLS protocol is implemented in Erlang, so only
the cryptographic algorithms come from the OpenSSL cryptolib.`

Author: Fixed

the cryptograpic algorithms comes from OpenSSL cryptolib.

> ## General Principals

Reviewer: ### 5.7 Heading typo: "Principals" → "Principles"

**Severity:** Minor
**WHY:** Section heading reads "General Principals" — should be "General Principles".
**WHAT:** Line 50.
**HOW:** Change to `## General Principles`.

> Not covered by bjorng's review.

Author: Fixed

> ## General Principals

Reviewer: ### 5.24 "General Principles" single-sentence subsections

**Severity:** Minor
**WHY:** The "General Principles" section (lines 50–70) has 5 `###`
subsections each containing exactly one sentence. The heading-to-content
ratio is ~1:1 — headings carry more visual weight than the content they
introduce. This breaks reading flow and is inconsistent with the rest
of the guide where sections have multi-paragraph content.
**WHAT:** Lines 50–70: subsections "Algorithms selection", "Post Quantum
Algorithms", "Legacy", "Embedded Systems", "NSS Key Logging".
**HOW:** Recommended: collapse to a bullet list with bold labels
(e.g. `- **Algorithm selection:** ...`). Each item becomes one bullet
instead of a heading + single sentence. Alternative: expand each
subsection with option references and examples (follow-up PR).

Author: This  section has been removed. It was a work in progress to try find the best structure.

### Algorithms selection
TLS protocol typically defines that it is the clients algorithm

> preference that should determine the selection. There exist a few

Reviewer: ### 5.45 Vague "a few server options"

**Severity:** Minor
**WHY:** Line 54–55: "There exist a few server options to instead
adhere to the servers preference" doesn't name the options. A hardening
guide should be actionable — the reader needs to know *which* options.
There are exactly two: `honor_cipher_order` and `honor_ecc_order`
(both default to `false`, both server-only). Related to AP-12 (missing
links) but distinct: AP-12 is about linking already-named options;
here the options aren't even mentioned.
**WHAT:** Lines 54–55.
**HOW:** Suggested rewrite:

> The TLS protocol typically specifies that the client's algorithm
> preference determines the selection. The server options
> [`honor_cipher_order`](`t:ssl:server_option_cert/0`) and
> [`honor_ecc_order`](`t:ssl:server_option_cert/0`) override this to
> use the server's preference instead.

Author: fixed

server options to instead adhere to the servers preference.

### Post Quantum Algorithms
Only supported in the latest protocol version TLS-1.3

> ### Legacy

Reviewer: ### 5.30 Security warnings missing admonition blocks

**Severity:** Minor
**WHY:** The guide has 4 sentences warning about unsafe/insecure
practices, all rendered as plain text. The guide already uses
`> #### Note {: .info }` blocks for informational notes (lines 46,
111) but has zero `> #### Warning {: .warning }` blocks. Security
warnings should be visually distinct. Not covered by bjorng.
**WHAT:**

| Line | Warning text | Recommended block |
|------|-------------|-------------------|
| 61–62 | "enabling anything documented as legacy will make your application less secure" | `{: .warning \
}` |
| 69–70 | NSS key logging "should not be enabled in production...defeats the purpose of TLS" | `{: .warni\
ng }` |
| 119 | "Do not use the same PSK across both TLS 1.2 and TLS 1.3...unsafe" | `{: .warning }` |
| 139–140 | verify_fun "at own risk and will weaken the security properties" | `{: .warning }` |

**HOW:** Wrap each in a warning admonition, e.g.:
```markdown
> #### Warning {: .warning }
> Do not use the same PSK across both TLS 1.2 and TLS 1.3,
> as it is considered unsafe.

Author: fixed

Be aware that enabling anything documented as legacy will make your
application less secure.

### Embedded Systems
As a trade-off for embedded systems cipher suites using AES-CCM_8 mode
instead of CCM mode are available.

### NSS Key Logging
This is a debugging functionality that should not be enabled in
production systems as it defeats the purpose of the TLS protocol.

## Protocol Versions
The protocol itself will prevent version downgrade when using TLS-1.3
and TLS-1.2 that are the default supported versions. All other TLS
versions still configurable are legacy versions. DTLS-1.2 over UDP can

> also be configured and has similar functionality as TLS-1.2 but will

Reviewer: ### 5.25 DTLS-1.2 "similar functionality" imprecise

**Severity:** Minor
**WHY:** Line 75 says DTLS-1.2 "has similar functionality as TLS-1.2".
Two issues: (1) grammar — "similar functionality as" should be
"similar functionality to"; (2) precision — DTLS-1.2 has the same
*security properties* (confidentiality, integrity, authentication) but
different *functionality* (no stream guarantees, different record
framing, cookie exchange for DoS mitigation, retransmission). The
sentence continuation about losing data partially addresses this, but
"similar functionality" is misleading.
**WHAT:** Line 75: `DTLS-1.2 over UDP can also be configured and has
similar functionality as TLS-1.2`
**HOW:** Replace "similar functionality" with "similar security
properties". bjorng's suggestion rewrites to "has functionality similar
to TLS-1.2" — fixes grammar but keeps the imprecise claim. Suggested:

> DTLS-1.2 over UDP can also be configured and provides similar
> security properties to TLS-1.2 but will always be less reliable
> and may lose application data just as plain UDP communication.

Author: fixed

always be less reliable and may loose application data just as plain
UDP communication. DTLS over other transports can possible be achieved
but not out of the box. DTLS-1.3 is currently not supported.

The protocol version TLS-1.3 is a major upgrade of the protocol
towards previous protocol versions and supporting both TLS-1.3 and
TLS-1.2 will require a configuration including options form both.

### Cipher Suites
TLS-1.3 does not share any cipher suites with previous protocol versions.
you need to make sure to include at least one cipher suite for both
versions to be able to support both versions in the `ciphers` option.

### Session Key Renewal
All options related to renegotiation are relevant to versions
previously to TLS-1.3 that replaces that mechanism with the key update
mechanism that only handles session key updates and hence is simpler
than the previous renegotiation mechanism.

> Client renegotiation can be disabled and is by default mitigate

Reviewer: ### 5.26 Client renegotiation: "mitigate" and vague rate-limit

**Severity:** Minor
**WHY:** Two issues on line 95–96: (1) "mitigate" → "mitigated" (passive
voice needs past participle); (2) "not allowing two renegotiations close
to each other" is vague — the actual mechanism is a 12-second cooldown
(`?WAIT_TO_ALLOW_RENEGOTIATION` in `ssl_connection.hrl:156`). Not
covered by bjorng's review.
**WHAT:** Lines 95–96: `Client renegotiation can be disabled and is by
default mitigate by the server not allowing two renegotiations close to
each other.`
**HOW:** Suggested rewrite:

> Client renegotiation can be disabled. By default, the server
> mitigates renegotiation abuse by enforcing a several seconds delay
> between renegotiations.

Author: Addressed!

by the server not allowing two renegotiations close to each other.

### Key Exchange Groups

> TLS-1.3 decouples key exchange algorithms form cipher suites. These

Reviewer: ### 5.28 Ambiguous "These" in key exchange paragraph

**Severity:** Minor
**WHY:** Line 99–100: "TLS-1.3 decouples key exchange algorithms from
cipher suites. These are configured using the `groups` option." —
"These" is ambiguous. The nearest noun is "cipher suites" but the
intended referent is "key exchange algorithms". bjorng's suggestion
(comment #14) fixes other issues in this paragraph but leaves "These"
unchanged.
**WHAT:** Line 100: `These are configured using the \`groups\` option.`
**HOW:** Replace "These" with the explicit subject:

> The key exchange algorithms are configured using the `groups` option.

Author: Is addressed now.

> are configured using the `groups` option. In TLS-1.2 in addition to

Reviewer: ### 5.27 Missing links to option and function documentation

**Severity:** Major
**WHY:** The guide references 10+ SSL options and functions but none
are linked to their documentation. The SSH hardening guide links every
option to its type spec or module anchor. A hardening guide's purpose
is to help users configure options — without clickable links, users
must manually search the `ssl` module docs. Not covered by bjorng.
**WHAT:** All option/function references in the guide:

| Line | Reference | Suggested link target |
|------|-----------|----------------------|
| 42, 88 | `ciphers` | `t:ssl:cipher_suites/0` |
| 44 | `ssl:filter_cipher_suites/2` | `ssl:filter_cipher_suites/2` |
| 101 | `groups` | `supported_groups` option in `t:ssl:common_option_tls13/0` |
| 102 | `eccs` | `t:ssl:common_option_pre_tls13/0` |
| 102 | `dh`, `dhfile` | server option types |
| 106, 112 | `signature_algs` | `t:ssl:signature_algs/0` |
| 106, 112 | `signature_algs_cert` | option doc |
| 128 | `certs_keys` | `t:ssl:cert_key_conf/0` |
| 134 | `verify_fun` | `t:ssl:common_option_cert/0` |

**HOW:** Add EEx-style doc links following the pattern used in the SSH
hardening guide, e.g.:
```markdown
[`ciphers`](`t:ssl:cipher_suites/0`)
[`ssl:filter_cipher_suites/2`](`ssl:filter_cipher_suites/2`)
[`groups`](`t:ssl:common_option_tls13/0`)
[`verify_fun`](`t:ssl:common_option_cert/0`)
```

Author: As linking is step two in process of coming up with the content.  Should be addressed now.

the cipher suites the options `eccs`, `dh`, `dhfile` exist to configure
the key exchange

### Signature Algorithms
The options `signature_algs` and `signature_algs_cert` can be used to
allow different signature algorithms for protocol messages and
certificates. If the latter is not specified it will default to
former.

> #### Note {: .info }
Currently the default will add `signature_algs_cert` that allows

> signature algorithms using sha1 as pseudo random function for

Reviewer: ### 5.15 Typo: "sha1" → "SHA-1"

**Severity:** Minor ⚡ bjorng
**WHAT:** Line 113: "sha1" → "SHA-1" (NIST standard name per FIPS
180-4). OTP source code uses the atom `sha` / `sha1`, but in
documentation prose the standard form SHA-1 should be used.

Author: Should be addressed now.

certificate signatures, which is not allowed by default for protocol
signatures.  Expected to be removed after year 2030 when such
certificates should have been phased out.

### Pre Shared Keys
Do not use the same PSK across both TLS 1.2 and TLS 1.3, as it is considered unsafe.

In TLS-1.2 PSK is supported by PSK cipher suites and generally lack forward secrecy.
In TLS-1.3  PSK can be use together with a key exchange algorithm
and then provides forward secrecy.

### Certificate and Keys

Use the `certs_keys` option to be able to configure more than one
possible certificate for the client or server and then the `tm` best
one that can be used with the peer will be chosen.

> ## Certificate Verification

Reviewer: ### 5.36 Missing "Certificate Verification" intro paragraph

**Severity:** Nit (non-blocking)
**WHY:** The `## Certificate Verification` section (line 132) jumps
directly into `### Verify Function` without an introductory paragraph.
Every other `##` section in the guide has at least one sentence of
context before its subsections. The SSH hardening guide's equivalent
section ("Verifying the remote daemon") has a multi-paragraph
introduction explaining the trust model.
**WHAT:** Line 132.
**HOW:** Add a brief intro, e.g.:

> Certificate verification is essential for establishing trust in TLS
> connections. The `ssl` application provides several mechanisms for
> customizing and strengthening certificate validation.

Author: Added

### Verify Function
The verify_fun option to TLS clients and servers lets you customize
the certificate path validation, that is validation the chain of
certificates that shall ensure the authenticity of the peer, and it's
purpose is to allow for application specific additional checks.  The
verify fun can however also be used to accept certain errors. Doing
the latter is at own risk and will weaken the security properties of
your application.

### Certificate Revocation Verification
Although this should be done it needs to be configured as the ssl
application does not know how the users application handles
this. There exists a basic default cache that can be externally
populated or rely on certificate HTTP distribution-points, a callback
API can also be use to access an external cache.

CRL check option best_effort is to be considered legacy.

The client can also support OCSP stapling if connecting to a server
that supports that. In TLS-1.2 only end entity certificate staple is supported.

> Client will fail if no staple is provided. Plain OCSP is not supported.

Reviewer: ### 5.35 "Client will fail" — imprecise hard-fail description

**Severity:** Minor
**WHY:** Line 155: "Client will fail if no staple is provided" is
imprecise. The actual behavior (verified in `ssl_handshake.erl:2379–2386`)
is that the connection fails with `{bad_cert, missing_ocsp_staple}`,
which can be intercepted by a custom `verify_fun` to implement fallback
revocation checking. The guide makes it sound like an unconditional
hard failure with no recourse.
**WHAT:** Line 155.
**HOW:** Suggested rewrite:

> If the server does not provide an OCSP staple, the connection will
> fail with `{bad_cert, missing_ocsp_staple}`. A custom `verify_fun`
> can intercept this to implement fallback revocation checking (e.g.
> direct CRL fetch).
>
> > #### Warning {: .warning }
> > Accepting `{bad_cert, missing_ocsp_staple}` without performing
> > alternative revocation checking is insecure, as it allows a MITM
> > attacker to suppress revocation information by omitting the staple.

This aligns with the documentation in `ssl.erl:1513–1521` and with
AP-15's recommendation to use warning admonitions for security advice.

Author: Add text along these lines.

> Client will fail if no staple is provided. Plain OCSP is not supported.

Reviewer: ### 5.42 "Plain OCSP" — ambiguous term

**Severity:** Minor
**WHY:** Line 156: "Plain OCSP is not supported" is ambiguous — "plain"
compared to what? The intended meaning is that the `ssl` application
does not perform direct OCSP queries (client contacting the OCSP
responder via the certificate's AIA extension); only OCSP stapling
(server-provided responses in the TLS handshake) is supported. A reader
unfamiliar with the distinction may not understand what is excluded.
**WHAT:** Line 156: `Plain OCSP is not supported.`
**HOW:** Change to:

> Direct OCSP queries by the client are not supported.

Author: Add note about this.

<!--
%CopyrightBegin%

SPDX-License-Identifier: Apache-2.0

Copyright Ericsson AB 2026-2026. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

%CopyrightEnd%
-->

# TLS Harding Guide

## Introduction

There can exist trade-offs between interoperability and security, and

> between resource consumption and security. If you makes such

Reviewer 2: ```suggestion
between resource consumption and security. If you make such
```

Author: fixed

trade-offs it is important that you make an informed decision and
understand the consequences.

> Erlang TLS implementation provided by the ssl application strives to

Reviewer 2: ```suggestion
Erlang's TLS implementation provided by the `ssl` application strives to
```

Author: fixed

have secure defaults, that said we will avoid changing defaults in

> none major releases.

Reviewer 2: ```suggestion
have secure defaults. That said, we will avoid changing defaults in
non-major releases.
```

Author: fixed

## Cryptographic Algorithms

> The available cryptographic algorithms depends on the version of

Reviewer 2: ```suggestion
The available cryptographic algorithms depend on the version of
```

Author: fixed

OpenSSL cryptolib that Erlang/OTP is built and linked with.

Configuration algorithm lists will be automatically filtered
except for `ciphers` that pre TLS-1.3 is combining several algorithms
and you need to call ssl:filter_cipher_suites/2

> #### Note {: .info }
Note that the TLS protocol is implemented in Erlang so only
the cryptography comes from OpenSSL cryptolib.

## General Principals

# Algorithms selection

TLS protocol typically defines that it is the clients algorithm
preference that should determine the selection. There exist a few

> server options to instead adhere to the servers preference.

Reviewer 2: ```suggestion
The TLS protocol typically specifies that the client's algorithm
preference determines the algorithm selection. However, there are a few
server options to instead adhere to the server's preference.
```

Author: Start is applied now,  the end has changed.

# Post Quantum Algorithms

> Only supported in the latest protocol version TLS-1.3

Reviewer 2: ```suggestion
Post-quantum algorithms are only supported in the latest protocol version TLS-1.3.
```

Author: Not relevant any more.

# Legacy

Be aware that enabling anything documented as legacy will make your
application less secure.

# Embedded Systems

As a trade-off for embedded systems cipher suites using AES-CCM_8 mode
instead of CCM mode are available.

# NSS Key Logging

> This is a debugging functionality that should not be enabled in

Reviewer 2: ```suggestion
NSS key logging is a debugging functionality that should not be enabled in
```

Author: Change has been applied

production systems as it defeats the purpose of the TLS protocol.

## Protocol Versions

The protocol itself will prevent version downgrade when using TLS-1.3
and TLS-1.2 that are the default supported versions. All other TLS
versions still configurable are legacy versions. DTLS-1.2 over UDP can
also be configured and has similar functionality as TLS-1.2 but will
always be less reliable and may loose application data just as plain
UDP communication. DTLS over other transports can possible be achieved

> but not out of the box. DTLS-1.3 is currently not supported.

Reviewer 2: ```suggestion
The protocol itself will prevent version downgrade when using TLS-1.3
and TLS-1.2, which are the default supported versions. All other TLS
versions still configurable are legacy versions. DTLS-1.2 over UDP can
also be configured and has functionality similar to TLS-1.2 but will
always be less reliable and may lose application data just like plain
UDP communication. DTLS over other transports can possibly be achieved
but not out of the box. DTLS-1.3 is currently not supported.
```

Reviewer 2: Did you see this comment?

Author: Applied. Now I have seen it GitHub auto collapsed a lot stuff comments when Kuba added his AI-generated comments.

The protocol version TLS-1.3 is a major upgrade of the protocol
towards previous protocol versions and supporting both TLS-1.3 and

> TLS-1.2 will require a configuration including options form both.

Reviewer 2: ```suggestion
The protocol version TLS-1.3 is a major upgrade
over previous versions, and supporting both TLS-1.3 and
TLS-1.2 will require a configuration including options from both.
```

Reviewer 2: Did you see this comment?  It should bee from now.

Author: Applied

Reviewer: #### 4.1.2 (AP-2/5.10) "form" → "from" Typo

**Severity:** Minor
**WHAT:** `ssl_hardening.md`, line 131: "options form both"
**HOW:** Change `form` to `from`.

# Cipher Suites

TLS-1.3 does not share any cipher suites with previous protocol versions.
you need to make sure to include at least one cipher suite for both

> versions to be able to support both versions in the `ciphers` option.

Reviewer 2: ```suggestion
You will need to include at least one cipher suite for both
versions to be able to support both in the `ciphers` option.
```

Author: Applied

# Session Key Renewal

All options related to renegotiation are relevant to versions
previously to TLS-1.3 that replaces that mechanism with the key update
mechanism that only handles session key updates and hence is simpler

> than the previous renegotiation mechanism.

Reviewer 2: I am not sure I fully understand this paragraph, but here is my best guess:

```suggestion
Renegotiation options only apply to versions
prior to TLS-1.3. In TLS-1.3, they have been replaced with the key update
mechanism,  which only handles session key updates and therefore is simpler
than the previous renegotiation mechanism.
```

Reviewer: double space in: `mechanism,  which`

should be gone

# Key Exchange Groups

TLS-1.3 decouples key exchange algorithms form cipher suites. These
are configured using the `groups` option. In TLS-1.2 in addition to
the cipher suites the options `eccs`, `dh`, `dhfile` exist to configure

> the key exchange

Reviewer 2: ```suggestion
TLS-1.3 decouples key exchange algorithms from cipher suites. These
are configured using the `groups` option. In TLS-1.2, in addition to
the cipher suites the options `eccs`, `dh`, and `dhfile` are available for configuring
the key exchange.
```

Author: Applied

# Signature Algorithms

The options `signature_algs` and `signature_algs_cert` can be used to
allow different signature algorithms for protocol messages and
certificates. If the latter is not specified it will default to

> former.

Reviewer 2: ```suggestion
certificates. If the `signature_algs_cert` option is not specified it will default to
use the algorithms requested by the `signature_algs` option.
```

Author: I do not see how this is confusing. But I rephrased it anyway!

> #### Note {: .info }

Currently the default will add `signature_algs_cert` that allows
signature algorithms using sha1 as pseudo random function for
certificate signatures, which is not allowed by default for protocol
signatures.  Expected to be removed after year 2030 when such

> certificates should have been phased out.

Reviewer 2: ```suggestion
Currently, the default includes `signature_algs_cert` that allows
signature algorithms using SHA1 as pseudo random function for
certificate signatures. This is not allowed by default for protocol
signatures. It is expected to be removed after 2030, when such
certificates should have been phased out.
```

Author: Adopted most of the comment and rephrased a little to keep it correct.

# Pre Shared Keys

TODO

# Certificate and Keys

Use the `certs_keys` option to be able to configure more than one
possible certificate for the client or server and then the `tm` best

> one that can be used with the peer will be chosen.

Reviewer 2: ```suggestion
Use the `certs_keys` option to be able to configure more than one
possible certificate for the client or server. The "best"
one compatible with the peer will be chosen.
```

Author: fixed

## Certificate Verification

# Verify Function

The verify_fun option to TLS clients and servers lets you customize
the certificate path validation, that is validation the chain of
certificates that shall ensure the authenticity of the peer, and it's
purpose is to allow for application specific additional checks.  The
verify fun can however also be used to accept certain errors. Doing
the latter is at own risk and will weaken the security properties of

> your application.

Reviewer 2: ```suggestion
The `verify_fun` option to TLS clients and servers lets you customize
the certificate path validation, that is, validation of the chain of
certificates that will ensure the authenticity of the peer, and its
purpose is to allow for application-specific additional checks. However,
the verify fun can also be used to accept certain errors. Doing
so is at your own risk and will weaken the security properties of
your application.
```

Author: fixed

# Certificate Revocation Verification

Although this should be done it needs to be configured as the ssl
application does not know how the users application handles
this. There exists a basic default cache that can be externally
populated or rely on certificate HTTP distribution-points, a callback

> API can also be use to access an external cache.

Reviewer 2: ```suggestion
Certificate revocation verification is important for security and should
not be skipped. However, it needs to be configured as the `ssl`
application does not know how the user's application handles
this. There is a basic default cache that can be externally
populated or rely on certificate HTTP distribution-points. There is
also a callback API that can be used to access an external cache.
```

Author: fixed

> CRL check option best_effort is to be considered legacy.

Reviewer 2: ```suggestion
TheCRL check option `best_effort` is to be considered legacy.
```

Author: Fixed

`OpenSSL cryptolib`.

### Algorithms selection
TLS protocol typically defines that it is the clients algorithm
preference that should determine the selection. There exist a few
server options to instead adhere to the servers preference.

### Signature Algorithms
The options `signature_algs` and `signature_algs_cert` can be used to
allow different signature algorithms for protocol messages and
certificates. If the latter is not specified it will default to
former.

> #### Note {: .info }
Currently using the default `signature_algs`
will add `signature_algs_cert` that allows signature algorithms using
sha1 as pseudo random function for certificate signatures, which is
not allowed by default for protocol signatures.  Expected to be
removed after year 2030 when such certificates should have been phased
out.

### Post Quantum Algorithms
Only supported in the latest protocol version TLS-1.3

## Protocol Versions
The protocol itself will prevent version downgrade when using TLS-1.3
and TLS-1.2 that are the default supported versions. All other TLS
versions still configurable are legacy versions. DTLS-1.2 over UDP can
also be configured and has similar functionality as TLS-1.2 but will

> always be less reliable and may loose application data just as plain

Reviewer 2: ```suggestion
always be less reliable and may lose application data just as plain
```

Author: Fixed

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

%CopyrightEnd%
-->

# TLS Hardening Guide
This guide describes how to harden TLS connections configured with
the ssl application. It covers cryptographic algorithm selection,
protocol version configuration, certificate verification, and
revocation checking.

> #### Note {: .info }
There can exist trade-offs between interoperability and security, and
between resource consumption and security. If you make such
trade-offs it is important that you make an informed decision and
understand the consequences.

Erlang's TLS implementation provided by the `ssl` application strives to
have secure defaults. That said, we will avoid changing defaults in
non-major releases.

> As a trade-off for embedded systems cipher suites using AES-CCM_8 mode

Reviewer: ### 5.12 (AP-39) AES-CCM_8 Sentence: Grammar and Precision

**Severity:** Minor
**WHY:** Line 39-40 reads:

> As a trade-off for embedded systems cipher suites using AES-CCM_8 mode
> instead of CCM mode are available.

Three problems:

1. **Missing comma** creates ambiguity — "embedded systems cipher
   suites" reads as a noun phrase. Should be: "for embedded systems**,**
   cipher suites..."

2. **"Trade-off" is unexplained.** CCM uses a 16-byte authentication
   tag; CCM-8 uses 8 bytes — weaker integrity (2⁶⁴ vs 2¹²⁸ forgery
   resistance) but saves 8 bytes per TLS record. The sentence doesn't
   say what's being traded for what.

3. **"Available" understates the situation.** CCM-8 suites are in the
   **default** cipher suite list (`tls_v1:exclusive_suites/1`, lines
   521, 528, 540) — both `TLS_AES_128_CCM_8_SHA256` (TLS-1.3) and
   `TLS_ECDHE_ECDSA_WITH_AES_*_CCM_8` (TLS-1.2). "Available" implies
   opt-in; they're actually enabled by default.

**WHAT:** `ssl_hardening.md`, lines 39-40.
**HOW:** Suggested rewrite:

> For embedded systems with constrained bandwidth, cipher suites using
> AES-CCM-8 (8-byte authentication tag) are included in the defaults
> alongside the full AES-CCM (16-byte tag). The shorter tag reduces
> per-record overhead at the cost of weaker integrity protection.

Author: Added rewrite of the rewrite.

instead of CCM mode are available.

> #### Warning {: .warning }
Be aware that enabling anything documented as legacy will make your
application less secure.

> #### Warning {: .warning }
NSS Key Logging is a debugging functionality that should not be enabled in
production systems as it defeats the purpose of the TLS protocol.

## Cryptographic Algorithms
The available cryptographic algorithms depend on the version of
OpenSSL cryptolib that Erlang/OTP is built and linked with.

> The configuration algorithm lists will be automatically filtered

Reviewer: ### 5.11 (AP-38) Cipher Suites Paragraph Confusing and Imprecise

**Severity:** Minor
**WHY:** Lines 55-58 read:

> The configuration algorithm lists will be automatically filtered
> except for cipher suites ([`ciphers`](`t:ssl:common_option/0`), option)
> that pre TLS-1.3 combines several algorithms
> and you will need to call `ssl:filter_cipher_suites/2`

Three problems:

1. **"Automatically filtered" is unexplained.** The reader has no idea
   what this means. The actual behavior (verified in source): when you
   configure `signature_algs`, `supported_groups`, or `eccs`, the `ssl`
   application filters the list against `crypto:supports()` at option
   processing time — algorithms your OpenSSL doesn't support are
   silently dropped (`tls_v1:signature_schemes/2` line 912,
   `tls_v1:groups/1` line 1180). This is not documented anywhere.

2. **The cipher suites exception is poorly explained.** The reason
   cipher suites are different: user-supplied cipher lists go through
   `binary_cipher_suites/2` which intersects against `all_suites/1` —
   but `all_suites` is **not** filtered through
   `ssl_cipher:filter_suites/1` (which checks `crypto:supports()`).
   Only the **default** cipher list is crypto-filtered
   (`default_binary_suites/2`, `ssl.erl` line 2762). So a user can
   configure cipher suites that reference algorithms their OpenSSL
   doesn't support, leading to runtime handshake failures.

3. **"Combines several algorithms" is vague.** The actual distinction:
   pre-TLS-1.3 cipher suites bundle 4 components (key exchange +
   authentication + cipher + MAC), so filtering requires checking all
   4 against crypto support — that's what
   `ssl:filter_cipher_suites/2` does via
   `ssl_cipher:crypto_support_filters/0`. In TLS-1.3, cipher suites
   only specify AEAD cipher + hash, and key exchange is separate.

**WHAT:** `ssl_hardening.md`, lines 55-58.
**HOW:** Suggested rewrite:

> When you configure algorithm options such as
> [`supported_groups`](`t:ssl:common_option_tls13/0`) or
> [`signature_algs`](`t:ssl:signature_algs/0`), the `ssl` application
> automatically removes any algorithms not supported by the underlying
> OpenSSL cryptolib. The exception is the
> [`ciphers`](`t:ssl:common_option/0`) option: user-supplied cipher
> suites are not automatically filtered against crypto support. In
> pre-TLS-1.3, each cipher suite combines key exchange, authentication,
> cipher, and MAC algorithms, so you should call
> [`ssl:filter_cipher_suites/2`](`ssl:filter_cipher_suites/2`) to
> remove suites that reference unsupported algorithms.

Author: Rephrased using suggestion.

except for cipher suites ([`ciphers`](`t:ssl:common_option/0`), option)
that pre TLS-1.3 combines several algorithms
and you will need to call `ssl:filter_cipher_suites/2`

> #### Note {: .info }
The TLS protocol is implemented in Erlang, so only the cryptographic
algorithms come from the `OpenSSL cryptolib`. This reduces the error
surface by excluding buffer overflows and other pointer related errors
that does not exist in Erlang programs.

### Algorithms selection

> TLS protocol typically defines that it is the clients algorithm

Reviewer: ### 5.2 (AP-33) Algorithm Selection Sentence

**Severity:** Minor
**WHY:** Lines 67-73 read:

> TLS protocol typically defines that it is the clients algorithm
> preference that should determine the selection. Server option
> `honor_cipher_order` and server option, for TLS-1.2 an legacy
> versions, `honor_ecc_order` currently exists to instead honer the
> servers preference of those algorithms.

Six problems:

1. **"honer" → "honor"** — not a word.
2. **"an legacy" → "and legacy"** (also AP-34).
3. **Subject-verb disagreement** — compound subject ("option X **and**
   option Y") requires "exist", not "exists".
4. **Garbled word order** — "currently exists to instead honer" splits
   the infinitive awkwardly; "currently exists" is filler.
5. **"of those algorithms" — vague referent.** No clear antecedent.
   `honor_cipher_order` controls cipher suite selection;
   `honor_ecc_order` controls ECC curve selection — they don't control
   the same "algorithms."
6. **Misplaced qualifier contradicts the code change.** The
   parenthetical "for TLS-1.2 an legacy versions" is meant to qualify
   only `honor_ecc_order`, but the sentence structure makes it read as
   qualifying both options. The PR's own code change (moving
   `honor_cipher_order` from `server_option_pre_tls13` to
   `server_option`) proves `honor_cipher_order` applies to ALL TLS
   versions including TLS-1.3. Verified: `honor_cipher_order` is used
   in `tls_server_connection_1_3.erl:398`; `honor_ecc_order` is absent
   from TLS 1.3 code paths.

**WHAT:** `ssl_hardening.md`, lines 67-73.
**HOW:** Suggested rewrite:

> The TLS protocol typically specifies that the client's algorithm
> preference determines the selection. The server option
> [`honor_cipher_order`](`t:ssl:server_option/0`) overrides this for
> cipher suite selection in all TLS versions. For TLS-1.2 and legacy
> versions, [`honor_ecc_order`](`t:ssl:server_option_pre_tls13/0`)
> does the same for ECC curve selection.

Author: Use the suggestion

preference that should determine the selection. Server option
[`honor_cipher_order`](`t:ssl:server_option/0`) and server option,
for TLS-1.2 an legacy versions,
[`honor_ecc_order`](`t:ssl:server_option_pre_tls13/0`) currently
exists to instead honer the servers
preference of those algorithms.

### Signature Algorithms

> The options [`signature_algs`](`t:ssl:signature_algs/0`) and

Reviewer: ### 4.13 (AP-41) Signature Algorithms Sentence

**Severity:** Minor
**WHY:** Lines 76-79 read:

> The options `signature_algs` and `signature_algs_cert` can be used
> to allow different signature algorithms for protocol messages and
> certificates. If the latter is not specified it will default to
> former.

Four problems:

1. **"default to former" — grammatically broken and technically
   misleading.** Missing articles ("the former"). More importantly,
   the Note block immediately below (lines 81-87) reveals the default
   `signature_algs_cert` is NOT identical to `signature_algs` — it's
   `signature_algs` **plus** legacy SHA-1 algorithms for certificate
   signatures. Confirmed in code: when `signature_algs_cert` is
   `undefined`, `ssl_certificate:verify_sign/2` (line 601) falls back
   to `signature_algs`, but the configured default adds SHA-1 on top.
   So it doesn't "default to the former" — it defaults to a superset.

2. **"allow" — imprecise verb.** The options don't "allow" anything —
   they *configure which* signature algorithms are acceptable.

3. **Missing comma.** "If the latter is not specified**,** it will..."

4. **"for protocol messages and certificates" — ambiguous scope.** A
   reader might think both options apply to both things. The actual
   split: `signature_algs` → TLS handshake protocol signatures
   (CertificateVerify, etc.); `signature_algs_cert` → certificate
   chain signatures.

**WHAT:** `ssl_hardening.md`, lines 76-79.
**HOW:** Suggested rewrite:

> The [`signature_algs`](`t:ssl:signature_algs/0`) option configures
> which signature algorithms are acceptable for TLS protocol messages.
> The [`signature_algs_cert`](`t:ssl:signature_algs/0`) option
> separately configures acceptable algorithms for certificate chain
> signatures. If `signature_algs_cert` is not specified, certificate
> signatures are verified against `signature_algs` (with additional
> legacy SHA-1 algorithms — see note below).

Author: Well no the AI is somewhat hallucinating. But rephrased using most of its suggestion.

> `signature_algs_cert` can be used to allow different signature

Reviewer: ### 5.3.1 (AP-12) Missing Links to Option and Function Documentation

**Severity:** Major (partially fixed)
**WHY:** 7 option/function references remain as plain backticks without
doc links. A hardening guide's purpose is to help users configure
options — clickable links are essential.

**Still missing links:**

| Line | Reference | Suggested link target |
|------|-----------|----------------------|
| 58 | `` `ssl:filter_cipher_suites/2` `` | `` [`ssl:filter_cipher_suites/2`](`ssl:filter_cipher_suites/2`) `` |
| 77 | `` `signature_algs_cert` `` | `` [`signature_algs_cert`](`t:ssl:signature_algs/0`) `` |
| 82 | `` `signature_algs` `` (Note block) | `` [`signature_algs`](`t:ssl:signature_algs/0`) `` |
| 83 | `` `signature_algs_cert` `` (Note block) | `` [`signature_algs_cert`](`t:ssl:signature_algs/0`) `` |
| 177 | `` `verify_fun` `` (Warning block) | `` [`verify_fun`](`t:ssl:common_option_cert/0`) `` |
| 179 | `` `verify` `` | `` [`verify`](`t:ssl:common_option_cert/0`) `` |
| 180 | `verify_none` (no backticks) | `` [`verify_none`](`t:ssl:common_option_cert/0`) `` |

Author: Added mor links. The construct  `ssl:filter_cipher_suites/2` will be linked.

algorithms for protocol messages and certificates. If the latter is
not specified it will default to former.

> #### Note {: .info }
Currently using the default `signature_algs`
will add `signature_algs_cert` that allows signature algorithms using
sha1 as pseudo random function for certificate signatures, which is
not allowed by default for protocol signatures.  Expected to be
removed after year 2030 when such certificates should have been phased
out.

> ### Post Quantum Algorithms

Reviewer: ### 4.14 (AP-42) PQC Section

**Severity:** Minor
**WHY:** Lines 89-107 (Post Quantum Algorithms) have 8 issues:

1. **"Quantum computers" should be "cryptographically relevant quantum
   computers" (CRQCs).** Current quantum computers can't break RSA or
   ECC. The threat is from future CRQCs with enough stable qubits for
   Shor's algorithm. NIST and the PQC community use "CRQC" to
   distinguish from today's NISQ devices. Without this qualifier, a
   reader could misinterpret the threat as imminent.

2. **"currently latest" is redundant** (line 92). Just "TLS-1.3".

3. **"break standards like RSA and ECC" — RSA and ECC are algorithms,
   not standards.** Standards are FIPS 186, RFC 8017, etc. Use
   "algorithms such as RSA and ECC."

4. **"will in general have properties that will increase" — vague and
   wordy** (lines 95-97). Which properties? Larger keys, larger
   signatures (SLH-DSA), larger ciphertexts (ML-KEM). Should be
   concrete.

5. **"battled tested" → "battle-tested"** (line 107). Missing hyphen,
   extra "d".

6. **"that is used to balance" — dangling modifier** (lines 105-106).
   "That" refers to `x25519mlkem768` but grammatically attaches to
   "default". Run-on sentence.

7. **No mention of "harvest now, decrypt later" threat model.** This
   is THE reason PQC matters for TLS today — adversaries record
   encrypted traffic now for future decryption with CRQCs. A hardening
   guide should state this to motivate enabling PQC before CRQCs exist.

8. **"For them to be used appropriate certificates and keys must be
   supplied too" — missing comma and unclear scope** (lines 101-102).
   This applies to ML-DSA/SLH-DSA (PQC certificates needed), not to
   ML-KEM key exchange (works with any cert type). The sentence doesn't
   distinguish.

**WHAT:** `ssl_hardening.md`, lines 89-107.
**HOW:** Suggested rewrite:

> Post-Quantum Cryptography (PQC) algorithms are designed to resist
> attacks from future cryptographically relevant quantum computers
> (CRQCs), which are expected to break algorithms such as RSA and ECC.
> PQC is only available in TLS-1.3.
>
> > #### Note {: .info }
> > Even though CRQCs do not yet exist, enabling PQC protects against
> > "harvest now, decrypt later" attacks where an adversary records
> > encrypted traffic today for future decryption.
>
> PQC algorithms generally require more computation and network
> bandwidth than classical algorithms. Key sizes, signatures, and
> ciphertexts are larger.
>
> For key exchange, several ML-KEM hybrid groups are supported. By
> default, only `x25519mlkem768` is offered — a hybrid that combines
> classical X25519 with ML-KEM-768 to balance current security with
> future quantum resistance, as the newer algorithms are less
> battle-tested.
>
> For authentication, the signature algorithms ML-DSA and SLH-DSA are
> supported. ML-DSA is preferred due to faster signing and smaller
> signatures. Using PQC signature algorithms requires appropriate PQC
> certificates and keys.

Author: Added suggestion

only the hybrid `x25519mlkem768` will be offered by default, that is
used to balance current security with future threats, as new algorithms
should be good but are less battled tested.

## Protocol Versions
The protocol itself will prevent version downgrade when using TLS-1.3
and TLS-1.2 that are the default supported versions. All other TLS
versions still configurable are legacy versions. DTLS-1.2 over UDP can
also be configured and has similar functionality as TLS-1.2 but will
always be less reliable and may lose application data just as plain
UDP communication. DTLS over other transports can possible be achieved
but not out of the box. DTLS-1.3 is currently not supported.

The protocol version TLS-1.3 is a major upgrade of the protocol
towards previous protocol versions and supporting both TLS-1.3 and
TLS-1.2 will require a configuration including options form both.

### Cipher Suites
TLS-1.3 does not share any cipher suites with previous protocol versions.
you need to make sure to include at least one cipher suite for both
versions to be able to support both versions in the
[`ciphers`](`t:ssl:common_option/0`) option.

> ### Session Key Renewal

Reviewer: ### 4.15 (AP-11) Session Key Renewal Section

**Severity:** Minor
**WHY:** Lines 128-135 read:

> All options related to renegotiation are relevant to versions
> previously to TLS-1.3 that replaces that mechanism with the key
> update mechanism that only handles session key updates and hence is
> simpler than the previous renegotiation mechanism.
>
> Client renegotiation can be disabled and is by default mitigate by
> the server not allowing two renegotiations close to each other.

Seven problems:

1. **"previously to" → "prior to"** (AP-2/5.12, bjorng #13).

2. **Run-on sentence** (line 128-132). 46 words, three clauses chained
   with "that" and "and". Tries to say: (a) renegotiation is pre-1.3,
   (b) TLS-1.3 replaced it with key update, (c) key update is simpler.

3. **"that replaces that mechanism" — double "that".** First is a
   relative pronoun (TLS-1.3), second is a demonstrative
   (renegotiation). Ambiguous.

4. **"only handles session key updates and hence is simpler" —
   misleading simplification.** Doesn't explain *why* it's simpler:
   renegotiation re-ran the full handshake mid-connection (new cipher
   suites, certificates, everything); key update only rotates keys.

5. **"mitigate" → "mitigated"** — passive voice requires past
   participle.

6. **"not allowing two renegotiations close to each other" — vague.**
   The actual mechanism: a 12-second cooldown timer
   (`?WAIT_TO_ALLOW_RENEGOTIATION = 12000` in
   `ssl_connection.hrl:156`, used in
   `tls_server_connection.erl:344`).

7. **"Client renegotiation can be disabled" — doesn't name the
   option.** The option is `client_renegotiation` (defaults to `true`,
   `ssl_config.erl:1297`). A hardening guide should name and link it.

**WHAT:** `ssl_hardening.md`, lines 128-135.
**HOW:** Suggested rewrite:

> Renegotiation options only apply to versions prior to TLS-1.3.
> TLS-1.3 replaces renegotiation with the key update mechanism, which
> only handles session key rotation and does not allow renegotiating
> cipher suites or certificates mid-connection.
>
> Client-initiated renegotiation can be disabled with the
> [`client_renegotiation`](`t:ssl:server_option_pre_tls13/0`) option.
> By default, the server mitigates renegotiation abuse by enforcing a
> 12-second delay between renegotiations.

Author: Rephrased from suggestion.

### Early Data
The security properties of Early Data introduced in TLS-1.3 are weaker
than other kinds of TLS data. See [Early Data in
TLS-1.3](using_ssl.html#early-data-in-tls-1-3) for more information of
what to consider.

## Certificate Handling

### Certificate and Keys
Use the [`certs_keys`](`t:ssl:common_option_cert/0`) option to be able
to configure more than one possible certificate for the client or
server. The `best` one compatible with the peer will be chosen.

### Verify Function
The [`verify_fun`](`t:ssl:common_option_cert/0`) option to TLS clients
and servers lets you customize the certificate path validation, that
is, validation of the chain of certificates that shall ensure the
authenticity of the peer, and its purpose is to allow for
application-specific additional checks.
> #### Warning {: .warning }
However, the `verify_fun` can also be used to accept certain
errors. Doing so is at your own risk and will weaken the security
properties of your application. Legacy client `verify` option
verify_none is implemented as a `verify_fun` that accepts all errors
and should only be used for testing and debugging purposes.

> ### Certificate Revocation Verification

Reviewer: ### 4.16 (AP-43) Certificate Revocation Section

**Severity:** Minor
**WHY:** Lines 183-198 (Certificate Revocation Verification) have 7
problems:

1. **"does not know how the user's application handles this" —
   confusing.** The `ssl` app doesn't enable CRL checking by default
   because it requires infrastructure. The sentence implies confusion;
   the reality is a deliberate design choice. Doesn't name the option:
   `crl_check` defaults to `false` (`ssl_config.erl:1499`).

2. **"basic default cache...externally populated or rely on..." —
   run-on and imprecise.** The cache is `ssl_crl_cache`. "Externally
   populated" = `ssl_crl_cache:insert/1`. "Rely on HTTP distribution
   points" = auto-fetch from certificate CRL DP URIs. Two different
   modes that should be explained separately.

3. **"callback API...access an external cache" — vague.** The API is
   `ssl_crl_cache_api` behaviour. "Access an external cache" undersells
   it — you implement the behaviour to provide CRLs from any source.
   Should name and link it.

4. **`best_effort` is "legacy" with no explanation of why.** Because
   it silently accepts certificates when revocation status can't be
   determined — an attacker blocking CRL/OCSP access gets a free pass.
   The PR's own `ssl.erl` change adds "must be considered insecure!"
   but the guide doesn't explain the risk.

5. **Missing backticks around `{bad_cert, missing_ocsp_staple}`**
   (line 197-198).

6. **No mention that `crl_check` defaults to `false`.** A hardening
   guide should state this and recommend enabling it.

7. **No mention of `crl_check` values `peer` vs `true`.** Four values
   exist: `false` (default), `best_effort` (legacy), `peer` (peer
   cert only), `true` (entire chain). Guide should recommend and
   explain the difference.

**WHAT:** `ssl_hardening.md`, lines 183-198.
**HOW:** Suggested rewrite:

> Certificate revocation verification is important for security.
> The [`crl_check`](`t:ssl:common_option_cert/0`) option is `false`
> by default and should be enabled. Use `true` to verify the entire
> certificate chain, or `peer` to verify only the peer certificate.
>
> > #### Warning {: .warning }
> > The `best_effort` value is legacy and insecure — it silently
> > accepts certificates when revocation status cannot be determined.
>
> The built-in CRL cache ([`ssl_crl_cache`](`m:ssl_crl_cache`)) can
> be pre-populated via `ssl_crl_cache:insert/1` or configured to
> fetch CRLs from HTTP distribution points in certificates. For
> custom CRL sources, implement the
> [`ssl_crl_cache_api`](`m:ssl_crl_cache_api`) behaviour.
>
> The client can request OCSP stapling from the server with the
> [`stapling`](`t:ssl:client_option_cert/0`) option (disabled by
> default). In TLS-1.2, only end-entity certificate stapling is
> supported. If the server does not provide a staple, the connection
> fails with `{bad_cert, missing_ocsp_staple}`.

Author: Rephrased with help of suggestion.

to remove suites that are not supported by linked `OpenSSL cryptolib`.

> #### Note {: .info }
The TLS protocol is implemented in Erlang, so only the cryptographic
algorithms come from the `OpenSSL cryptolib`. This reduces the error
surface by excluding buffer overflows and other pointer related errors
that does not exist in Erlang programs.

### Algorithms selection
The TLS protocol typically specifies that the client's algorithm
preference determines the selection. The server option
[`honor_cipher_order`](`t:ssl:server_option/0`) overrides this for
cipher suite selection in all TLS versions. For TLS-1.2 and legacy
versions,[`honor_cipher_order`](`t:ssl:server_option/0`)
does the same for ECC curve selection.

### Signature Algorithms
The [`signature_algs`](`t:ssl:signature_algs/0`) option configures
which signature algorithms are acceptable for TLS protocol messages.
The [`signature_algs_cert`](`t:ssl:signature_algs/0`), option
separately configures acceptable algorithms for certificate chain
signatures.

If [`signature_algs`](`t:ssl:signature_algs/0`) is specified but
not [`signature_algs_cert`](`t:ssl:signature_algs/0`), the latter

> will default to the former, however if none of them are specified

Reviewer 2: I find this phrase confusing: "the latter will default to the former". I would suggest that you try to get rid of "latter" and "former". I don't feel I have the technical knowledge to do it myself.

```suggestion
not [`signature_algs_cert`](`t:ssl:signature_algs/0`), the latter
will default to the former; however, if none of them are specified
```

Author: I did rephrase it !

Reviewer: I agree with Bjorn. This is unnecessary prose in document which should be direct and actionable. As non native person, I always wonder which one is the latter and which one is the former ;-)

(CRQCs), which are expected to break algorithms such as RSA and ECC.
PQC is only available in TLS-1.3.

Note {: .info }
Even though CRQCs do not yet exist, enabling PQC protects against
"harvest now, decrypt later" attacks where an adversary records
encrypted traffic today for future decryption.

PQC algorithms generally require more computation and network
bandwidth than classical algorithms. Key sizes, signatures, and
ciphertexts are larger.

For key exchange, several ML-KEM hybrid groups are supported. By
default, only x25519mlkem768 is offered — a hybrid that combines
classical X25519 with ML-KEM-768 to balance current security with
future quantum resistance, as the newer algorithms are less
battle-tested.

For authentication, the signature algorithms ML-DSA and SLH-DSA are
supported. ML-DSA is preferred due to faster signing and smaller
signatures. Using PQC signature algorithms requires appropriate PQC
certificates and keys.

## Protocol Versions
The protocol itself will prevent version downgrade when using TLS-1.3
and TLS-1.2 that are the default supported versions. All other TLS
versions still configurable are legacy versions. DTLS-1.2 over UDP can

> also be configured and has similar functionality as TLS-1.2 but will

Reviewer 2: ```suggestion
also be configured and has functionality similar to TLS-1.2 but will
```

Author: fixed

> always be less reliable and may lose application data just as plain

Reviewer 2: ```suggestion
always be less reliable and may lose application data just like plain
```

Author: fixed

> UDP communication. DTLS over other transports can possible be achieved

Reviewer 2: ```suggestion
UDP communication. DTLS over other transports can possibly be achieved
```

Author: fixed

but not out of the box. DTLS-1.3 is currently not supported.

The protocol version TLS-1.3 is a major upgrade of the protocol
towards previous protocol versions and supporting both TLS-1.3 and
TLS-1.2 will require a configuration including options form both.

### Cipher Suites
TLS-1.3 does not share any cipher suites with previous protocol versions.
you need to make sure to include at least one cipher suite for both
versions to be able to support both versions in the
[`ciphers`](`t:ssl:common_option/0`) option.

### Session Key Renewal
Renegotiation options only apply to versions prior to TLS-1.3.
TLS-1.3 replaces renegotiation with the key update mechanism, which
only handles session key rotation and does not allow renegotiating
cipher suites or certificates mid-connection.

A server can disable client-initiated renegotiation with the
[`client_renegotiation`](`t:ssl:server_option_pre_tls13/0`) option to

> avoid possible dos-attacks. By default, the server mitigates

Reviewer 2: ```suggestion
avoid possible [DoS-attacks](https://en.wikipedia.org/wiki/Denial-of-service_attack).
By default, the server mitigates
```

Author: fixed

<!--
%CopyrightBegin%

SPDX-License-Identifier: Apache-2.0

Copyright Ericsson AB 2026-2026. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

%CopyrightEnd%
-->

> # Transport Layer Security (TLS) Hardening Guide

Reviewer: #### 4.1.12 (AP-50) Title Truncation in Sidebar

**Severity:** Minor
**WHY:** The title "Transport Layer Security (TLS) Hardening Guide"
(49 chars) is too long for the ExDoc left sidebar. It renders as
"Transport Layer Security (TLS) Har..." in the navigation menu at
https://erlang.org/github-pr/11019/lib/ssl-11.5.4/doc/html/ssl_hardening.html

No other OTP guide expands "TLS" in its title. The SSH hardening
guide is simply "Hardening". Other SSL guides use "TLS" without
expansion ("TLS/DTLS Protocol Overview", "Erlang Distribution over
TLS"). The `secure_coding.md` cross-reference already says "TLS
Hardening Guide".

**WHAT:** `ssl_hardening.md`, lines 23-24.
**HOW:**
1. Title: `# TLS Hardening Guide` — matches other SSL guide title
   lengths (20-28 chars) that render without truncation.
2. First sentence: expand TLS on first use in body text:
   "This guide describes how to harden **Transport Layer Security
   (TLS)** connections configured with the `ssl` application."
   This follows the guide's own convention (AES, AEAD, PQC, CRQC,
   PSK, UDP, DTLS, PFS, OCSP, CRL are all expanded on first use).

Author: fixed

This guide describes how to harden TLS connections configured with
the ssl application. It covers cryptographic algorithm selection,
protocol version configuration, certificate verification, and
revocation checking.

> #### Note {: .info }
There can exist trade-offs between interoperability and security, and
between resource consumption and security. If you make such
trade-offs it is important that you make an informed decision and
understand the consequences.

Erlang's TLS implementation provided by the `ssl` application strives to
have secure defaults. That said, we will avoid changing defaults in
non-major releases.

For embedded systems with constrained bandwidth, cipher suites using
Advanced Encryption Standard (AES) with counter cipher block chaining
message authentication code, that is AES-CCM-8 (8-byte authentication
tag) are included in the defaults further down the priority list than

> the full AES-CCM (16-byte tag). The shorter tag reduces

Reviewer: #### 4.1.8 (AP-39) AES-CCM-8 Paragraph

**Severity:** Minor (partially fixed — 2 of 3 fixed)

**Still open:** Missing period at end of paragraph (line 43).

Author: fixed

per-TLS-record-layer overhead at the cost of weaker integrity

> protection

Reviewer: #### 4.1.14 (AP-52) CCM-8 Hardening Recommendation

**Severity:** Minor
**WHY:** The AES-CCM-8 paragraph (lines 38-45) describes the
bandwidth/security trade-off but doesn't quantify the weakness or
recommend action. A hardening guide should tell users what to do.

Key facts the current text omits:
- CCM-8 provides 64-bit integrity vs 128-bit for all other defaults
- OpenSSL excludes CCM-8 from its default cipher list (security
  level 2 requires ≥112-bit; see Appendix A)
- No other major TLS implementation includes CCM-8 in defaults

**WHAT:** `ssl_hardening.md`, lines 38-45.
**HOW:** After the existing paragraph, add a recommendation:

```markdown
> #### Note {: .info }
> AES-CCM-8's 8-byte tag provides 64-bit integrity protection,
> compared to 128-bit for AES-GCM and AES-CCM. If you do not
> operate in a bandwidth-constrained environment, consider removing
> CCM-8 cipher suites using
> [`ssl:filter_cipher_suites/2`](`ssl:filter_cipher_suites/2`):
>
> ```erlang
> Suites = ssl:filter_cipher_suites(ssl:cipher_suites(default, 'tlsv1.3'),
>            [{cipher, fun(aes_128_ccm_8) -> false; (_) -> true end}]).
> ```
```

Author: Changed default to not include these. And updated guide considering suggestion.

> #### Warning {: .warning }
Be aware that enabling anything documented as legacy will make your
application less secure.

> #### Warning {: .warning }
Network Security Services Key Log Format (NSS Key Logging)
is a debugging functionality that should not be enabled in
production systems as it defeats the purpose of the TLS protocol.

## Cryptographic Algorithms
The available cryptographic algorithms depend on the version of
OpenSSL cryptolib that Erlang/OTP is built and linked with.

When you configure algorithm options such as
[`supported_groups`](`t:ssl:common_option_tls13/0`), or
[`signature_algs`](`t:ssl:common_option/0`), the ssl application
automatically removes any algorithms not supported by the underlying
OpenSSL cryptolib. The exception is the
[`ciphers`](`t:ssl:common_option/0`) option: user-supplied cipher
suites are not automatically filtered against crypto support, this is
due to the more complex composition of algorithms in cipher suites

> pre TLS-1.3. In this case use `ssl:filter_cipher_suites/2`

Reviewer: #### 4.1.10 (AP-12) Missing Doc Links (8 references)

**Severity:** Minor

| Line | Reference | Suggested link target |
|------|-----------|----------------------|
| 67 | `` `ssl:filter_cipher_suites/2` `` | `` [`ssl:filter_cipher_suites/2`](`ssl:filter_cipher_suites/2`) `` |
| 93 | `` `signature_algs` `` (2nd paragraph) | already linked above, but plain here |
| 96 | `` signature_algs `` (no backticks) | `` [`signature_algs`](`t:ssl:signature_algs/0`) `` |
| 100 | `` `signature_algs` `` (Note block) | `` [`signature_algs`](`t:ssl:signature_algs/0`) `` |
| 207 | `early_data` (no backticks) | `` [`early_data`](`t:ssl:server_option/0`) `` |
| 253 | `stapling option` (no backticks) | `` [`stapling`](`t:ssl:client_option_cert/0`) `` |
| 258 | `verify_fun` (no backticks) | `` [`verify_fun`](`t:ssl:common_option_cert/0`) `` |
| 247 | `best_effort` (no backticks) | `` `best_effort` `` |

Author: Line 67 is AI hallucination, it will be linked.  Rest is duplicate comments!

Reviewer: sorry.

to remove suites that are not supported by linked `OpenSSL cryptolib`.

> #### Note {: .info }
The TLS protocol is implemented in Erlang, and therefore only the cryptographic
functions come from the `OpenSSL cryptolib`. This reduces the error
surface by excluding buffer overflows and other pointer related errors
that does not exist in Erlang programs.

### Algorithms selection
The TLS protocol typically specifies that the client's algorithm
preference determines the algorithm selection. The server option

> [`honor_cipher_order`](`t:ssl:server_option/0`) overrides this for

Reviewer: #### 4.1.1 (AP-44) `honor_cipher_order` vs `honor_ecc_order`

**Severity:** Major
**WHY:** Line 80 reads:

> For TLS-1.2 and legacy versions,[`honor_cipher_order`](`t:ssl:server_option/0`)
> does the same for ECC curve selection.

This should be `honor_ecc_order`, not `honor_cipher_order`. The option
that controls ECC curve selection is `honor_ecc_order`
(`server_option_pre_tls13`). `honor_cipher_order` controls cipher
suite selection (already stated on line 78). Copy-paste error — the
previous version correctly had `honor_ecc_order`.

**WHAT:** `ssl_hardening.md`, line 80.
**HOW:** Change `honor_cipher_order` to `honor_ecc_order` and update
the link target:
```
versions, [`honor_ecc_order`](`t:ssl:server_option_pre_tls13/0`)
```

Author: Hoops, fixed.

cipher suite selection in all TLS versions. For TLS-1.2 and legacy

> versions,[`honor_cipher_order`](`t:ssl:server_option/0`)

Reviewer: #### 4.1.12 (AP-48) Missing Space After Comma

**Severity:** Nit
**WHAT:** Line 80: "versions,[`honor" → "versions, [`honor..."

Author: fixed

does the same for ECC curve selection.

### Signature Algorithms
The [`signature_algs`](`t:ssl:signature_algs/0`) option configures
which signature algorithms are acceptable for TLS protocol messages.

> The [`signature_algs_cert`](`t:ssl:signature_algs/0`), option

Reviewer: #### 4.1.6 (AP-41) Signature Algorithms Paragraph

**Severity:** Minor (partially fixed — 2 of 4 original issues fixed)

**Still open:**
1. **Double spaces** on line 94: "of  [`signature_algs`]...  will"
2. **Stray comma** on line 88: `` [`signature_algs_cert`](...)**,** option ``
3. **Run-on sentence** (lines 92-97): two clauses joined with
   "however" without semicolon or period.
4. **`signature_algs` without backticks** on line 96.
5. **"none of them are specified" is ambiguous.**

**HOW:** Suggested rewrite for lines 92-97:

> If only [`signature_algs`](`t:ssl:signature_algs/0`) is specified,
> its value is also used for
> [`signature_algs_cert`](`t:ssl:signature_algs/0`). When neither
> option is specified, the defaults include additional legacy SHA-1
> algorithms for certificate signatures (see note below).

Author: Fixed small adjustments and rephrased so AI should stop hallucinating.

separately configures acceptable algorithms for certificate chain
signatures.

If [`signature_algs`](`t:ssl:signature_algs/0`) is specified but
not [`signature_algs_cert`](`t:ssl:signature_algs/0`), the value
of  [`signature_algs`](`t:ssl:signature_algs/0`)  will implicitly be used
for [`signature_algs_cert`](`t:ssl:signature_algs/0`), however if none of them are specified

> certificate signatures are verified against signature_algs (with additional

Reviewer: #### 4.1.5 (AP-47) Missing Backticks and Links

**Severity:** Minor

| Line | Text | Should be |
|------|------|-----------|
| 96 | `signature_algs` (plain text) | `` [`signature_algs`](`t:ssl:signature_algs/0`) `` |
| 100 | `` `signature_algs` `` (no link, Note block) | `` [`signature_algs`](`t:ssl:signature_algs/0`) `` |
| 201 | `early_data` (plain text) | `` [`early_data`](`t:ssl:server_option/0`) `` |
| 248 | `best_effort` (plain text) | `` `best_effort` `` |
| 258 | `stapling option` (plain text) | `` [`stapling`](`t:ssl:client_option_cert/0`) option `` |
| 263 | `verify_fun` (plain text) | `` [`verify_fun`](`t:ssl:common_option_cert/0`) `` |
| 267 | `{bad_cert, missing_ocsp_staple}` (plain in Warning) | `` `{bad_cert, missing_ocsp_staple}` `` |
| 264 | "e.g.  direct" (double space) | "e.g. direct" |

Author: fixed

TLS 1.3 significantly improves session handling over TLS 1.2 by
reducing the handshake from two round-trips (2-RTT) to one (1-RTT).
It mandates forward secrecy, encrypts more of the handshake, and
introduces 0-RTT resumption for previously visited sites.

The protocol itself will prevent version downgrade when using TLS-1.3
and TLS-1.2, which are the default supported versions. All other TLS
versions still configurable are legacy versions. Datagram transport layer security,
DTLS-1.2 over User Datagram Protocol (UDP) can also be configured and has
functionality similar to TLS-1.2 but will always be less reliable and may lose
application data just like plain UDP communication. DTLS over other
transports can possibly be achieved but not out of the box. DTLS-1.3
is currently not supported.

### Cipher Suites
TLS-1.3 does not share any cipher suites with previous protocol versions.
You will need to include at least one cipher suite for both
TLS-1.3 and TLS-1.2 to be able to support both versions in the
[`ciphers`](`t:ssl:common_option/0`) option.

TLS 1.3 uses only modern Authenticated Encryption with Associated Data (AEAD)

> cipher suites, so does ssl applications TLS-1.2 default.

Reviewer: #### 4.1.3 (AP-45) Grammar: "so does ssl applications TLS-1.2 default"

**Severity:** Minor
**WHY:** This phrase appears twice:
- Line 156: "TLS 1.3 uses only modern AEAD cipher suites, so does ssl
  applications TLS-1.2 default."
- Line 178: "which provides mandatory Perfect Forward Secrecy (PFS),
  so does ssl applications TLS-1.2 default."

Both are grammatically broken. "So does" requires a subject that
matches the verb.

**HOW:** Change both to:
"as does the `ssl` application's TLS-1.2 default."

Author: fixed.

### Key Exchange Groups
TLS-1.3 decouples key exchange algorithms from cipher suites. The key
exchange algorithms are configured using the
[`supported_groups`](`t:ssl:common_option_tls13/0`) option. In TLS-1.2
in addition to the cipher suites the options
[`eccs`](`t:ssl:common_option_pre_tls13/0`),
[`dh`](`t:ssl:server_option_pre_tls13/0`),
[`dhfile`](`t:ssl:server_option_pre_tls13/0`), are available for
configuring the key exchange.

TLS 1.3 removes static RSA, supporting only ephemeral Diffie-Hellman,
which provides mandatory Perfect Forward Secrecy (PFS),
so does ssl applications TLS-1.2 default.

### Pre-Shared Keys
> #### Warning {: .warning }
Do not use the same Pre-Shared Key (PSK) across both TLS 1.2 and TLS 1.3, as it is
considered unsafe.

In TLS-1.2, PSK is supported by PSK cipher suites, which generally
lack forward secrecy unless combined with DHE or ECDHE key exchange.
In TLS-1.3, PSK are most commonly used together with a key exchange algorithm,
providing forward secrecy, although a PSK only option still exists.

### Early Data
Early data (0-RTT) in TLS-1.3 is not forward secret and is
vulnerable to replay attacks. The server option

> early_data is disabled by default.

Reviewer: #### 4.1.7 (AP-31) Early Data Section

**Severity:** Minor (partially fixed — 2 of 4 fixed)

**Still open:**
1. `early_data` on line 201 — no backticks, no doc link.
2. Link still uses `.html` on line 208: `using_ssl.html#...` → `using_ssl.md#...`
3. Duplicate "See" sentences (lines 207-208). (AP-49)

Author: Fixed

> #### Warning {: .warning }
Only enable early data if your application can safely handle
replayed requests (that is for instance, idempotent HTTP methods).

See Early Data in TLS-1.3 for mitigation strategies and examples.
See [Early Data in TLS-1.3](using_ssl.html#early-data-in-tls-1-3).

## Authenticity
Certificate verification is essential for establishing trust in TLS
connections. The `ssl` application provides several mechanisms for
customizing and strengthening certificate validation.

### Certificate and Keys
Use the [`certs_keys`](`t:ssl:common_option_cert/0`) option to be able
to configure more than one possible certificate for the client or
server. The `best` one compatible with the peer will be chosen.

### Verify Function

> The [`verify_fun`](`t:ssl:common_option_cert/0`) option to TLS clients

Reviewer: #### 4.1.19 (AP-57) Run-on Opening Sentence

**Severity:** Nit
**WHY:** Lines 221-225 is a single sentence with two subordinate
clauses ("that is..." and "and its purpose..."). Hard to parse.

**WHAT:** `ssl_hardening.md`, lines 221-225.
**HOW:** Split into two sentences:

> The [`verify_fun`](`t:ssl:common_option_cert/0`) option lets you
> customize certificate path validation for both TLS clients and
> servers. Use it to add application-specific checks beyond the
> standard chain verification.

Author: fixed

and servers lets you customize the certificate path validation, that
is, validation of the chain of certificates that shall ensure the
authenticity of the peer, and its purpose is to allow for
application-specific additional checks.

> > #### Warning {: .warning }

Reviewer: #### 4.1.15 (AP-53) Verify Function: Warning Block Overloaded

**Severity:** Minor
**WHY:** The Warning block (lines 226-238) mixes three distinct topics:
1. `verify_fun` can be used to weaken security
2. `verify_none` on client is dangerous
3. `verify_none` on server is default (informational)

Item 3 is not a warning — it's an explanation of why the server
default differs. It should be outside the warning block, possibly
in a Note or plain text.

**WHAT:** `ssl_hardening.md`, lines 226-238.
**HOW:** Split the warning block. Keep items 1-2 in the Warning.
Move item 3 (server-side `verify_none` explanation) to plain text
or a Note block after the warning.

Author: fixed

However, the [`verify_fun`](`t:ssl:common_option_cert/0`) can also be
used to accept certain errors. Doing so is at your own risk and will

> weaken the security properties of your application. Legacy client

Reviewer: #### 4.1.16 (AP-54) `verify` and `fail_if_no_peer_cert` Not Legacy

**Severity:** Minor
**WHY:** Line 229: "Legacy client `verify` option" — `verify` is a
current, active option in both `client_option_cert/0` and
`server_option_cert/0`. It is not deprecated or legacy. What changed
is the *default value* on the client side: `verify_none` →
`verify_peer` in OTP 26 (commit `bb3603db84`, 2023-03-13, by Ingela).

Similarly, line 236-237: "the server legacy option
`fail_if_no_peer_cert`" — this option is not legacy. It's a current
`server_option_cert` option with no deprecation notice.

**WHAT:** `ssl_hardening.md`, lines 229 and 236-237.
**HOW:** Change "Legacy client `verify` option set to the value
`verify_none`" to "Setting the client `verify` option to
`verify_none`". Remove "legacy" from `fail_if_no_peer_cert`.

Author: So this is for a major release so lets fix the code and the documentation.

> [`verify`](`t:ssl:common_option_cert/0`) option set to the value

Reviewer: #### 4.1.21 (AP-59) Wrong Link Target for `verify`

**Severity:** Nit
**WHY:** Line 230: `[`verify`](`t:ssl:common_option_cert/0`)` — but
`verify` is not in `common_option_cert`. It's in
`client_option_cert/0` and `server_option_cert/0`. Since this
sentence discusses the client, link to `t:ssl:client_option_cert/0`.

**WHAT:** `ssl_hardening.md`, line 230.
**HOW:** Change link target to `t:ssl:client_option_cert/0`.

Author: fixed

> `verify_none` is implemented as a

Reviewer: #### 4.1.20 (AP-58) Implementation Detail in Guide

**Severity:** Nit
**WHY:** Lines 231-232: "`verify_none` is implemented as a
`verify_fun` that accepts all errors" — this exposes an internal
implementation detail. A hardening guide reader needs to know the
*effect* (disables all certificate verification), not the *how*
(internally a verify_fun).

**WHAT:** `ssl_hardening.md`, lines 231-232.
**HOW:** Rephrase to focus on effect: "Setting `verify_none` disables
all certificate verification and should only be used for testing."

Author: Rephrased to shift focus.

[`verify_fun`](`t:ssl:common_option_cert/0`) that accepts all errors
and should only be used for testing and debugging purposes. On the
server side however, `verify_none` is default as client certification
is an optional feature of the protocol and it needs configuring. When
client certification is configured the server legacy option

> [`fail_if_no_peer_cert`](`t:ssl:server_option_cert/0`) defaults to

Reviewer: #### 4.1.17 (AP-55) `fail_if_no_peer_cert` Conditional Default

**Severity:** Minor
**WHY:** Line 237-238: "defaults to true" — only when `verify` is
`verify_peer` on the server. Source: `ssl_config.erl:708`:

```erlang
DefFailNoPeer = Role =:= server andalso Verify =:= verify_peer,
```

The text says "When client certification is configured" which is
vague. Should explicitly state the condition.

**WHAT:** `ssl_hardening.md`, lines 237-238.
**HOW:** Change to: "defaults to `true` when `verify` is set to
`verify_peer`".

Author: Made it super clear!

true.

> 

Reviewer: #### 4.1.18 (AP-56) Missing Hardening Recommendations

**Severity:** Minor
**WHY:** The Verify Function section describes capabilities but
doesn't recommend actions. A hardening guide should include concrete
advice:

- Always use `verify_peer` on the client side (default since OTP 26)
- Avoid accepting `{bad_cert, _}` errors in `verify_fun` unless
  you have a documented, specific reason
- When enabling client certification on the server, ensure
  `fail_if_no_peer_cert` is `true`

**WHAT:** `ssl_hardening.md`, Verify Function section.
**HOW:** Add a brief hardening recommendation paragraph after the
warning block.

Author: fixed

### Certificate Revocation Verification
Certificate revocation verification is important for security and
needs to be performed. However it requires some application specific
knowledge to set it up correctly, hence the option
[`crl_check`](`t:ssl:common_option_cert/0`) default to
`false`. Use true to verify the entire certificate chain, or peer to
verify only the peer certificate.
> #### Warning {: .warning }
The best_effort value is legacy and insecure — it silently
accepts certificates when revocation status cannot be determined.

The built-in Certificate Revocation List (CRL) cache can
be pre-populated via `ssl_crl_cache:insert/1` or configured to
fetch CRLs from HTTP distribution points in certificates.
For custom CRL sources, implement the
[`ssl_crl_cache_api behaviour`](`m:ssl_crl_cache_api`).

The client can request Online Certificate Status Protocol (OCSP)
stapling from the server with the stapling option (disabled by

> default). In TLS-1.2, only end-entity certificate stapling is

Reviewer: #### 4.1.13 (AP-51) TLS 1.3 Multi-Stapling Not Mentioned

**Severity:** Nit
**WHY:** The Certificate Revocation section (line 253) states:

> In TLS-1.2, only end-entity certificate stapling is supported.

This correctly describes the TLS 1.2 limitation but doesn't mention
the TLS 1.3 improvement. TLS 1.3 (RFC 8446 §4.4.2.1) supports
multi-stapling — each `CertificateEntry` can carry its own OCSP
response, enabling staple verification for the entire certificate
chain. The OTP `ssl` code supports this: `split_cert_entries/4`
(`tls_handshake_1_3.erl:822`) stores per-cert OCSP responses in
`CertExt`, and `verify_cert_extensions/4`
(`ssl_certificate.erl:601`) verifies each one via
`public_key:pkix_ocsp_validate/5`.

A hardening guide could note this as a security advantage of TLS 1.3.

**WHAT:** `ssl_hardening.md`, after line 255.
**HOW:** Add a sentence such as:
"In TLS-1.3, the server can provide OCSP staples for each
certificate in the chain (multi-stapling), enabling revocation
verification of intermediate certificates as well."

Note: multi-stapling has no dedicated test case in
`openssl_stapling_SUITE`. OpenSSL ≤3.5.x does not support
multi-stapling (`chainidx != 0` guard in
`extensions_srvr.c:1755` skips intermediate certs). However,
**OpenSSL 4.0.0** added multi-stapling support (CHANGES.md:
"Added support for TLS 1.3 OCSP multi-stapling for server certs"
by Michael Krueger, Martin Rauch). New options:
- `s_server -status_all` — provides OCSP for entire chain
- `s_server -status_file` — can now be given multiple times
- `s_client -ocsp_check_all` — checks all certs in chain

Once OTP builds against OpenSSL 4.0.0, the `openssl_stapling_SUITE`
could be extended to test multi-stapling using `-status_all`.

Author: Added comment to reflect this. Extended test support is for an other PR.

supported. If the server does not provide a staple, the connection
by default fails with `{bad_cert, missing_ocsp_staple}`.

A custom verify_fun can intercept this to implement fallback
revocation checking (e.g.  direct CRL fetch).

> #### Warning {: .warning }
Accepting {bad_cert, missing_ocsp_staple} without performing
alternative revocation checking is insecure, as it allows a MITM
attacker to suppress revocation information by omitting the staple.

> > #### Note {: .note }

Reviewer: #### 4.1.22 (AP-60) Note Admonition: `{: .note }` Renders as Plain Blockquote

**Severity:** Minor
**WHY:** The last admonition block (line 269) uses `{: .note }` which
is not a standard ExDoc admonition type. ExDoc recognizes `.info`,
`.warning`, `.error`, and `.neutral`. Using `.note` causes ExDoc to
render it as a plain `<blockquote>` with a bare `<h4 class="note">`
instead of a styled `<section class="admonition info">`.

The content displays inside a rectangle (blockquote styling), but
it lacks the colored border, icon, and consistent styling that the
Warning and Note blocks earlier in the guide have. Those use
`{: .info }` and render as proper admonition boxes.

Rendered HTML comparison:
- Other Notes: `<section role="note" class="admonition info">` ✅
- This block: `<blockquote><h4 class="note">Note</h4>` — plain blockquote

**WHAT:** `ssl_hardening.md`, line 269.
**HOW:** Change `{: .note }` to `{: .info }` for consistent styling.

Author: fixed

The ssl application does not perform direct OCSP queries (client contacting the OCSP

> responder via the certificate's Authority Information Access (AIA) extension), only

Reviewer: ### 4.1.4 (AP-46) "only only" Duplicate Word

**Severity:** Minor
**WHY:** Line 270:

> only only OCSP stapling(server-provided responses in the TLS handshake) is supported.

Duplicate "only". Also missing space before parenthesis.

**HOW:** Remove duplicate "only", add space: "stapling (".

Author: fixed

you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

%CopyrightEnd%
-->

# TLS Hardening Guide
This guide describes how to harden Transport Layer Security (TLS)
connections configured with the ssl application. It covers
cryptographic algorithm selection, protocol version configuration,
certificate verification, and revocation checking.

> #### Note {: .info }
There can exist trade-offs between
interoperability and security, and between resource consumption and
security. If you make such trade-offs it is important that you make an
informed decision and understand the consequences. Also this guide
will always reflect the status of defaults for the release where it
is published, unless otherwise specifically stated. Hence using older

> releases may need more configuration to achive the same result.

Reviewer: #### 4.1.5 (AP-65) Typo "achive" → "achieve"

**Severity:** Minor
**WHY:** `ssl_hardening.md` line 36:
```
releases may need more configuration to achive the same result.
```

**WHAT:** `ssl_hardening.md`, line 36.
**HOW:** Change `achive` to `achieve`.

Author: fixed

Advanced Encryption Standard (AES) with counter cipher block chaining
message authentication code, that is AES-CCM-8 (8-byte authentication
tag) can be configured and provides 64-bit integrity instead of
128-bit.

> #### Warning {: .warning }
Be aware that enabling anything documented as legacy will make your
application less secure.

> #### Warning {: .warning }
Network Security Services Key Log Format (NSS Key Logging)
is a debugging functionality that should not be enabled in
production systems as it defeats the purpose of the TLS protocol.

## Cryptographic Algorithms
The available cryptographic algorithms depend on the version of
OpenSSL cryptolib that Erlang/OTP is built and linked with.

When you configure algorithm options such as
[`supported_groups`](`t:ssl:common_option_tls13/0`), or
[`signature_algs`](`t:ssl:common_option/0`), the ssl application
automatically removes any algorithms not supported by the underlying
OpenSSL cryptolib. The exception is the
[`ciphers`](`t:ssl:common_option/0`) option: user-supplied cipher
suites are not automatically filtered against crypto support, this is
due to the more complex composition of algorithms in cipher suites

> pre TLS-1.3. In this case use `ssl:filter_cipher_suites/2`

Reviewer: #### 4.1.10 (AP-12) Missing Doc Links

**Severity:** Minor
Remaining unlinked references:

| Line | Reference | Suggested link |
|------|-----------|----------------|
| 69 | `ssl:filter_cipher_suites/2` | `[`ssl:filter_cipher_suites/2`](`ssl:filter_cipher_suites/2`)` |

Author: I though AI should be able to learn ! The function reference with arity will be render a link!

Reviewer: I wasn't aware. seems to work indeed!

### Certificate and Keys
Use the [`certs_keys`](`t:ssl:common_option_cert/0`) option to be able
to configure more than one possible certificate for the client or
server. The `best` one compatible with the peer will be chosen.

### Verify Function
The [`verify_fun`](`t:ssl:common_option_cert/0`) option lets you
customize certificate path validation for both TLS clients and
servers. Use it to add application-specific checks beyond the
standard chain verification.

> #### Warning {: .warning }
However, the [`verify_fun`](`t:ssl:common_option_cert/0`) can also be
used to accept certain errors. Doing so is at your own risk and will
weaken the security properties of your application. Legacy client
[`verify`](`t:ssl:client_option_cert/0`) option set to the value
`verify_none` skips all certificate verification errors using
a special [`verify_fun`](`t:ssl:common_option_cert/0`)
and should only be used for testing and debugging purposes.

> #### Note {: .info }
On the server side, `verify_none` is default as client certification
is an optional feature of the protocol and it needs configuring. When
client certification is configured, setting

> [`verify`](`t:ssl:common_option_cert/0`) option to `verify_peer`, the

Reviewer: #### 4.1.6 (AP-66) Wrong `verify` Link Target (Line 244)

**Severity:** Minor
**WHY:** Line 244: `[`verify`](`t:ssl:common_option_cert/0`)` — but
`verify` is no longer in `common_option_cert`. In this PR, `verify`
was moved to `client_option_legacy` and `server_option_cert`. Since
line 244 discusses the server setting `verify` to `verify_peer`, the
link should target `t:ssl:server_option_cert/0`.

**WHAT:** `ssl_hardening.md`, line 244.
**HOW:** Change link to `t:ssl:server_option_cert/0`.

Author: fixed

server legacy option

> [`fail_if_no_peer_cert`](`t:ssl:server_option_cert/0`) defaults to

Reviewer: #### 4.1.7 (AP-67) `fail_if_no_peer_cert` Link Target Stale

**Severity:** Minor
**WHY:** Lines 246 and 254 link `fail_if_no_peer_cert` to
`t:ssl:server_option_cert/0`, but this PR moved it to
`server_option_legacy`. The link target should be updated.

**WHAT:** `ssl_hardening.md`, lines 246 and 254.
**HOW:** Change link target to `t:ssl:server_option_legacy/0`.

Author: fixed

true.

To summarize:

> - Always use verify_peer on the client side (default since OTP 26)

Reviewer: #### 4.1.10 (AP-12) Missing Doc Links

**Severity:** Minor
Remaining unlinked references:

| Line | Reference | Suggested link |
|------|-----------|----------------|
| 250 | `verify_peer` (plain text) | `` `verify_peer` `` |

Author: fixed

- Avoid accepting `{bad_cert, _}` errors in [`verify_fun`](`t:ssl:common_option_cert/0`) unless
  there is some edge-case workaround that you have confidence in excluding.
- When enabling client certification on the server, ensure

> [`fail_if_no_peer_cert`](`t:ssl:server_option_cert/0`) is true (default since OTP 26).

Reviewer: #### 4.1.7 (AP-67) `fail_if_no_peer_cert` Link Target — Partially Fixed

**Severity:** Minor
**WHY:** Line 247 now correctly links to `t:ssl:server_option_legacy/0`
✅. However, line 255 in the "To summarize" list still links to
`t:ssl:server_option_cert/0`:
```
[`fail_if_no_peer_cert`](`t:ssl:server_option_cert/0`) is true
```
Since `fail_if_no_peer_cert` was moved to `server_option_legacy` in
this PR, this link target is stale.

**WHAT:** `ssl_hardening.md`, line 255.
**HOW:** Change `t:ssl:server_option_cert/0` to
`t:ssl:server_option_legacy/0`.

Author: fixed

> #### Warning {: .warning }
The `best_effort` value is legacy and insecure — it silently
accepts certificates when revocation status cannot be determined.

The built-in Certificate Revocation List (CRL) cache can
be pre-populated via `ssl_crl_cache:insert/1` or configured to
fetch CRLs from HTTP distribution points in certificates.
For custom CRL sources, implement the
[`ssl_crl_cache_api behaviour`](`m:ssl_crl_cache_api`).

The client can request Online Certificate Status Protocol (OCSP)
stapling from the server with the
[`stapling`](`t:ssl:client_option_cert/0`) option (disabled by
default). In TLS-1.2, only end-entity certificate stapling is
supported as opposed to TLS-1.3. If the server does not provide a
staple, the connection by default fails with `{bad_cert,
missing_ocsp_staple}`.

A custom [`verify_fun`](`t:ssl:common_option_cert/0`) can intercept
this to implement fallback revocation checking (that is, direct CRL
fetch).

> #### Warning {: .warning }
Accepting `{bad_cert, missing_ocsp_staple}` without performing
alternative revocation checking is insecure, as it allows a MITM
attacker to suppress revocation information by omitting the staple.

> > #### Note {: .info } The ssl application does not perform direct

Reviewer: this does not render properly

#### 4.1.8 (AP-68) Admonition Header Formatting (Line 291)

**Severity:** Minor
**WHY:** Line 291 merges the admonition header with body text:
```
> #### Note {: .info } The ssl application does not perform direct
```
All other admonitions in the file have `{: .info }` or `{: .warning }`
on their own line, with body text starting on the next line. This
inconsistency may cause ExDoc rendering issues (the title text
"Note" and the body text run together).

**WHAT:** `ssl_hardening.md`, line 291.
**HOW:** Add a newline after `{: .info }`:
```
> #### Note {: .info }
The ssl application does not perform direct
```

Author: fixed

OCSP queries (client contacting the OCSP responder via the
certificate's Authority Information Access (AIA) extension), only OCSP
stapling (server-provided responses in the TLS handshake) is supported
to be verified by the client. The Erlang server has no stapling support.

> Also OCSP support seems to be on the decline dropped by for instance

Reviewer: #### 4.1.9 (AP-69) OCSP Decline Sentence Grammar (Line 296)

**Severity:** Nit
**WHY:** Line 296:
```
Also OCSP support seems to be on the decline dropped by for instance
```
Missing commas make this hard to parse. Should read:
"Also, OCSP support seems to be on the decline, dropped by, for
instance, `Let's Encrypt`." or restructured as:
"OCSP support also seems to be declining — for instance, `Let's
Encrypt` has dropped it."

**WHAT:** `ssl_hardening.md`, line 296.
**HOW:** Add commas or restructure the sentence.

Author: fixed
